// salesforce-alm/dist/commands/schema/schema_sobject_describe.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
const Messages = require("../../lib/force-cli-messages");
module.exports = function (_execCommand) {
    return {
        command: 'schema:sobject:describe',
        description: Messages.get('SchemaSObjectDescribeDescription'),
        longDescription: Messages.get('SchemaSObjectDescribeLongDescription'),
        help: Messages.get('SchemaSObjectDescribeHelp'),
        requiresProject: false,
        args: [],
        flags: [{
                name: 'sobjecttype',
                char: 's',
                hasValue: true,
                required: true,
                hidden: false,
                type: 'string',
                description: Messages.get('SchemaSObjectDescribeObjectNameDescription'),
                longDescription: Messages.get('SchemaSObjectDescribeObjectNameLongDescription')
            }],
        supportsTargetUsername: true,
        run(context) {
            return __awaiter(this, void 0, void 0, function* () {
                const SchemaSObjectDescribeCommand = require('../../lib/schema/schemaSObjectDescribeCommand').SchemaSObjectDescribeCommand;
                return yield _execCommand(new SchemaSObjectDescribeCommand(), context);
            });
        }
    };
};
