// salesforce-alm/dist/commands/apex/apex_execute.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
const Messages = require("../../lib/force-cli-messages");
module.exports = function (_execCommand) {
    return {
        command: 'apex:execute',
        description: Messages.get('ApexExecDescription'),
        longDescription: Messages.get('ApexExecLongDescription'),
        help: Messages.get('ApexExecHelp'),
        requiresProject: false,
        args: [],
        flags: [{
                name: 'apexcodefile',
                char: 'f',
                hasValue: true,
                required: false,
                hidden: false,
                type: 'file',
                description: Messages.get('ApexExecFilePathDescription'),
                longDescription: Messages.get('ApexExecFilePathLongDescription')
            }],
        supportsTargetUsername: true,
        run(context) {
            return __awaiter(this, void 0, void 0, function* () {
                const ApexExecuteCommand = require('../../lib/apex/apexExecuteCommand').ApexExecuteCommand;
                return yield _execCommand(new ApexExecuteCommand(), context);
            });
        }
    };
};
