// salesforce-alm/dist/commands/apex/apex_log_get.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
const Messages = require("../../lib/force-cli-messages");
module.exports = function (_execCommand) {
    return {
        command: 'apex:log:get',
        description: Messages.get('ApexLogGetDescription'),
        longDescription: Messages.get('ApexLogGetLongDescription'),
        help: Messages.get('ApexLogGetHelp'),
        requiresProject: false,
        args: [],
        flags: [
            {
                name: 'logid',
                char: 'i',
                hasValue: true,
                required: true,
                hidden: false,
                type: 'id',
                description: Messages.get('ApexLogGetIdDescription'),
                longDescription: Messages.get('ApexLogGetIdLongDescription')
            },
        ],
        supportsTargetUsername: true,
        run(context) {
            return __awaiter(this, void 0, void 0, function* () {
                const ApexLogGetCommand = require('../../lib/apex/apexLogGetCommand').ApexLogGetCommand;
                return yield _execCommand(new ApexLogGetCommand(), context);
            });
        }
    };
};
