// salesforce-alm/dist/commands/package1/package1_version_display.js
/*
 * Copyright (c) 2017, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root  or https://opensource.org/licenses/BSD-3-Clause
 */
/* eslint global-require: 0 */
'use strict';
const path = require('path');
const messages = require(path.join(__dirname, '..', '..', 'lib', 'messages'))();
module.exports = function (_execCommand) {
    return {
        description: messages.getMessage('package1VersionDisplayCommandCliDescription'),
        longDescription: messages.getMessage('package1VersionDisplayCommandLongDescription'),
        help: messages.getMessage('package1VersionDisplayCommandCliHelp'),
        supportsTargetUsername: true,
        flags: [
            {
                name: 'packageversionid',
                char: 'i',
                description: messages.getMessage('package1VersionDisplayCommandPackageId'),
                longDescription: messages.getMessage('package1VersionDisplayCommandPackageIdLong'),
                type: 'id',
                hasValue: true,
                required: true
            }
        ],
        /**
         * the heroku cli hook
         * @param context - the cli context
         * @returns {Promise}
         */
        run(context) {
            const Package1VersionDisplayCommand = require(path.join(__dirname, '..', '..', 'lib', 'package1', 'package1VersionDisplayCommand'));
            return _execCommand(new Package1VersionDisplayCommand(context.org), context);
        }
    };
};
