// salesforce-alm/dist/commands/package1/package1_version_list.js
/*
 * Copyright (c) 2017, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root  or https://opensource.org/licenses/BSD-3-Clause
 */
/* eslint global-require: 0 */
'use strict';
const path = require('path');
const messages = require(path.join(__dirname, '..', '..', 'lib', 'messages'))();
module.exports = function (_execCommand) {
    return {
        description: messages.getMessage('package1VersionListCommandCliDescription'),
        longDescription: messages.getMessage('package1VersionListCommandLongDescription'),
        help: messages.getMessage('package1VersionListCommandCliHelp'),
        supportsTargetUsername: true,
        flags: [
            {
                name: 'packageid',
                char: 'i',
                description: messages.getMessage('package1VersionListCommandPackageId'),
                longDescription: messages.getMessage('package1VersionListCommandPackageIdLong'),
                type: 'id',
                hasValue: true,
                required: false
            }
        ],
        /**
         * the heroku cli hook
         * @param context - the cli context
         * @returns {Promise}
         */
        run(context) {
            const Package1VersionListCommand = require(path.join(__dirname, '..', '..', 'lib', 'package1', 'package1VersionListCommand'));
            return _execCommand(new Package1VersionListCommand(context.org), context);
        }
    };
};
