// salesforce-alm/dist/commands/package2/package2_version_create.js
/*
 * Copyright (c) 2016, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root  or https://opensource.org/licenses/BSD-3-Clause
 */
/* eslint global-require: 0 */
'use strict';
const path = require('path');
const messages = require(path.join(__dirname, '..', '..', 'lib', 'messages'))();
const consts = require(path.join(__dirname, '..', '..', 'lib', 'constants'));
module.exports = function (_execCommand) {
    return {
        description: messages.getMessage('cliDescription', [], 'package2_version_create'),
        longDescription: messages.getMessage('cliLongDescription', [], 'package2_version_create'),
        help: messages.getMessage('help', [], 'package2_version_create'),
        orgType: consts.DEFAULT_DEV_HUB_USERNAME,
        supportsTargetDevHubUsername: true,
        flags: [
            {
                name: 'package2id',
                char: 'i',
                description: messages.getMessage('package2Id', [], 'package2_version_create'),
                longDescription: messages.getMessage('longPackage2Id', [], 'package2_version_create'),
                hasValue: true,
                required: false,
                type: 'id'
            },
            {
                name: 'directory',
                char: 'd',
                description: messages.getMessage('dir', [], 'package2_version_create'),
                longDescription: messages.getMessage('longDir', [], 'package2_version_create'),
                hasValue: true,
                required: false,
                type: 'directory'
            },
            {
                name: 'branch',
                char: 'b',
                description: messages.getMessage('branch', [], 'package2_version_create'),
                longDescription: messages.getMessage('longBranch', [], 'package2_version_create'),
                hasValue: true,
                required: false,
                type: 'string'
            },
            {
                name: 'tag',
                char: 't',
                description: messages.getMessage('tag', [], 'package2_version_create'),
                longDescription: messages.getMessage('longTag', [], 'package2_version_create'),
                hasValue: true,
                required: false,
                type: 'string'
            },
            {
                name: 'installationkey',
                char: 'k',
                description: messages.getMessage('key', [], 'package2_version_create'),
                longDescription: messages.getMessage('longKey', [], 'package2_version_create'),
                hasValue: true,
                required: false,
                type: 'string'
            },
            {
                name: 'preserve',
                char: 'p',
                description: messages.getMessage('preserve', [], 'package2_version_create'),
                longDescription: messages.getMessage('longPreserve', [], 'package2_version_create'),
                hasValue: false,
                required: false,
                hidden: true,
                type: 'boolean'
            },
            {
                name: 'validateschema',
                char: 'j',
                description: messages.getMessage('validateschema', [], 'package2_version_create'),
                longDescription: messages.getMessage('longValidateschema', [], 'package2_version_create'),
                hasValue: false,
                required: false,
                hidden: true,
                type: 'boolean'
            },
            {
                name: 'wait',
                char: 'w',
                description: messages.getMessage('wait', [], 'package2_version_create'),
                longDescription: messages.getMessage('longWait', [], 'package2_version_create'),
                hasValue: true,
                required: false,
                type: 'minutes',
                default: '0'
            }
        ],
        /**
         * the heroku cli hook
         * @param context - the cli context
         * @returns {Promise}
         */
        run(context) {
            const Package2VersionCreateCommand = require(path.join('..', '..', 'lib', 'package2', 'package2VersionCreateCommand'));
            return _execCommand(new Package2VersionCreateCommand(), context);
        }
    };
};
