// salesforce-alm/dist/commands/package2/package2_list.js
/*
 * Copyright (c) 2016, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root  or https://opensource.org/licenses/BSD-3-Clause
 */
/* eslint global-require: 0 */
'use strict';
const path = require('path');
const messages = require(path.join(__dirname, '..', '..', 'lib', 'messages'))();
const consts = require(path.join(__dirname, '..', '..', 'lib', 'constants'));
module.exports = function (_execCommand) {
    return {
        description: messages.getMessage('cliDescription', [], 'package2_list'),
        longDescription: messages.getMessage('cliLongDescription', [], 'package2_list'),
        help: messages.getMessage('help', [], 'package2_list'),
        requiresProject: false,
        orgType: consts.DEFAULT_DEV_HUB_USERNAME,
        supportsTargetDevHubUsername: true,
        flags: [],
        /**
         * the heroku cli hook
         * @param context - the cli context
         * @returns {Promise}
         */
        run(context) {
            const Package2ListCommand = require(path.join('..', '..', 'lib', 'package2', 'package2ListCommand'));
            return _execCommand(new Package2ListCommand(), context);
        }
    };
};
