// salesforce-alm/dist/commands/package2/package2_version_create_get.js
/*
 * Copyright (c) 2016, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root  or https://opensource.org/licenses/BSD-3-Clause
 */
/* eslint global-require: 0 */
'use strict';
const path = require('path');
const messages = require(path.join(__dirname, '..', '..', 'lib', 'messages'))();
const consts = require(path.join(__dirname, '..', '..', 'lib', 'constants'));
module.exports = function (_execCommand) {
    return {
        description: messages.getMessage('cliDescription', [], 'package2_version_create_get'),
        longDescription: messages.getMessage('cliLongDescription', [], 'package2_version_create_get'),
        help: messages.getMessage('help', [], 'package2_version_create_get'),
        requiresProject: false,
        orgType: consts.DEFAULT_DEV_HUB_USERNAME,
        supportsTargetDevHubUsername: true,
        flags: [{
                name: 'package2createrequestid',
                char: 'i',
                description: messages.getMessage('requestId', [], 'package2_version_create_get'),
                longDescription: messages.getMessage('requestIdLong', [], 'package2_version_create_get'),
                hasValue: true,
                required: true,
                type: 'id'
            }],
        /**
         * the heroku cli hook
         * @param context - the cli context
         * @returns {Promise}
         */
        run(context) {
            const Package2VersionCreateRequestGetCommand = require(path.join('..', '..', 'lib', 'package2', 'package2VersionCreateRequestGetCommand'));
            return _execCommand(new Package2VersionCreateRequestGetCommand(), context);
        }
    };
};
