// salesforce-alm/dist/commands/package2/package2_version_update.js
/*
 * Copyright (c) 2016, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root  or https://opensource.org/licenses/BSD-3-Clause
 */
/* eslint global-require: 0 */
'use strict';
const path = require('path');
const messages = require(path.join(__dirname, '..', '..', 'lib', 'messages'))();
const consts = require(path.join(__dirname, '..', '..', 'lib', 'constants'));
module.exports = function (_execCommand) {
    return {
        description: messages.getMessage('cliDescription', [], 'package2_version_update'),
        longDescription: messages.getMessage('cliLongDescription', [], 'package2_version_update'),
        help: messages.getMessage('help', [], 'package2_version_update'),
        orgType: consts.DEFAULT_DEV_HUB_USERNAME,
        supportsTargetDevHubUsername: true,
        flags: [
            {
                name: 'package2versionid',
                char: 'i',
                description: messages.getMessage('id', [], 'package2_version_update'),
                longDescription: messages.getMessage('idLong', [], 'package2_version_update'),
                hasValue: true,
                required: true,
                type: 'id'
            },
            {
                name: 'name',
                char: 'n',
                description: messages.getMessage('name', [], 'package2_version_update'),
                longDescription: messages.getMessage('nameLong', [], 'package2_version_update'),
                hasValue: true,
                required: false,
                type: 'string'
            },
            {
                name: 'description',
                char: 'd',
                description: messages.getMessage('description', [], 'package2_version_update'),
                longDescription: messages.getMessage('descriptionLong', [], 'package2_version_update'),
                hasValue: true,
                required: false,
                type: 'string'
            },
            {
                name: 'branch',
                char: 'b',
                description: messages.getMessage('branch', [], 'package2_version_update'),
                longDescription: messages.getMessage('branchLong', [], 'package2_version_update'),
                hasValue: true,
                required: false,
                type: 'string'
            },
            {
                name: 'tag',
                char: 't',
                description: messages.getMessage('tag', [], 'package2_version_update'),
                longDescription: messages.getMessage('tagLong', [], 'package2_version_update'),
                hasValue: true,
                required: false,
                type: 'string'
            },
            {
                name: 'installationkey',
                char: 'k',
                description: messages.getMessage('key', [], 'package2_version_update'),
                longDescription: messages.getMessage('longKey', [], 'package2_version_update'),
                hasValue: true,
                required: false,
                type: 'string'
            },
            {
                name: 'setasreleased',
                char: 's',
                description: messages.getMessage('setasreleased', [], 'package2_version_update'),
                longDescription: messages.getMessage('setasreleasedLong', [], 'package2_version_update'),
                hasValue: false,
                required: false,
                type: 'flag'
            }
        ],
        /**
         * the heroku cli hook
         * @param context - the cli context
         * @returns {Promise}
         */
        run(context) {
            const Package2VersionUpdateCommand = require(path.join('..', '..', 'lib', 'package2', 'package2VersionUpdateCommand'));
            return _execCommand(new Package2VersionUpdateCommand(), context);
        }
    };
};
