// salesforce-alm/dist/commands/package2/package2_version_create_list.js
/*
 * Copyright (c) 2016, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root  or https://opensource.org/licenses/BSD-3-Clause
 */
/* eslint global-require: 0 */
'use strict';
const path = require('path');
const messages = require(path.join(__dirname, '..', '..', 'lib', 'messages'))();
const consts = require(path.join(__dirname, '..', '..', 'lib', 'constants'));
module.exports = function (_execCommand) {
    return {
        description: messages.getMessage('cliDescription', [], 'package2_version_create_list'),
        longDescription: messages.getMessage('cliLongDescription', [], 'package2_version_create_list'),
        help: messages.getMessage('help', [], 'package2_version_create_list'),
        requiresProject: false,
        orgType: consts.DEFAULT_DEV_HUB_USERNAME,
        supportsTargetDevHubUsername: true,
        flags: [{
                name: 'createdlastdays',
                char: 'c',
                description: messages.getMessage('createdLastDaysDescription', [], 'package2'),
                longDescription: messages.getMessage('createdLastDaysLongDescription', [], 'package2'),
                hasValue: true,
                required: false,
                type: 'number'
            },
            {
                name: 'status',
                char: 's',
                description: messages.getMessage('statusDescription', [], 'package2_version_create_list'),
                longDescription: messages.getMessage('statusLongDescription', [], 'package2_version_create_list'),
                hasValue: true,
                required: false,
                type: 'string',
                values: ['Queued', 'InProgress', 'Success', 'Error']
            }
        ],
        /**
         * the heroku cli hook
         * @param context - the cli context
         * @returns {Promise}
         */
        run(context) {
            const Package2VersionCreateRequestListCommand = require(path.join('..', '..', 'lib', 'package2', 'package2VersionCreateRequestListCommand'));
            return _execCommand(new Package2VersionCreateRequestListCommand(), context);
        }
    };
};
