// salesforce-alm/dist/commands/lightning/lightning_test_install.js
/*
 * Copyright (c) 2016, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root  or https://opensource.org/licenses/BSD-3-Clause
 */
/* eslint global-require: 0 */
'use strict';
const path = require('path');
const messages = require(path.join(__dirname, '..', '..', 'lib', 'messages'))();
module.exports = function (_execCommand) {
    return {
        description: messages.getMessage('lightningTestInstallCommandDescription', [], 'lightning_test'),
        longDescription: messages.getMessage('lightningTestInstallCommandDescriptionLong', [], 'lightning_test'),
        help: messages.getMessage('lightningTestInstallCommandHelp', [], 'lightning_test'),
        requiresProject: false,
        supportsTargetUsername: true,
        flags: [{
                name: 'wait',
                char: 'w',
                description: messages.getMessage('lightningTestInstallCommandParamWait', [], 'lightning_test'),
                longDescription: messages.getMessage('lightningTestInstallCommandParamWaitLong', [], 'lightning_test'),
                hasValue: true,
                required: false,
                type: 'minutes',
                default: 2
            }, {
                name: 'releaseversion',
                char: 'r',
                description: messages.getMessage('lightningTestInstallCommandParamVersion', [], 'lightning_test'),
                longDescription: messages.getMessage('lightningTestInstallCommandParamVersionLong', [], 'lightning_test'),
                hasValue: true,
                required: false,
                type: 'string',
                default: 'latest'
            }, {
                name: 'packagetype',
                char: 't',
                description: messages.getMessage('lightningTestInstallCommandParamType', [], 'lightning_test'),
                longDescription: messages.getMessage('lightningTestInstallCommandParamTypeLong', [], 'lightning_test'),
                hasValue: true,
                required: false,
                type: 'string',
                values: ['jasmine', 'mocha', 'full'],
                default: 'full'
            }],
        run(context) {
            const LightningTestInstallCommand = require(path.join('..', '..', 'lib', 'lightning', 'lightningTestInstallCommand'));
            return _execCommand(new LightningTestInstallCommand(), context);
        }
    };
};
