// salesforce-alm/dist/commands/limits/limits_api_display.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
const Messages = require("../../lib/force-cli-messages");
module.exports = function (_execCommand) {
    return {
        command: 'limits:api:display',
        description: Messages.get('ApiLimitsDescription'),
        longDescription: Messages.get('ApiLimitsLongDescription'),
        help: Messages.get('ApiLimitsHelp'),
        requiresProject: false,
        args: [],
        flags: [],
        supportsTargetUsername: true,
        run(context) {
            return __awaiter(this, void 0, void 0, function* () {
                const ApiLimitsCommand = require('../../lib/limits/apiLimitsCommand').ApiLimitsCommand;
                return yield _execCommand(new ApiLimitsCommand(), context);
            });
        }
    };
};
