// salesforce-alm/dist/commands/project/project_upgrade.js
/*
 * Copyright (c) 2016, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root  or https://opensource.org/licenses/BSD-3-Clause
 */
/* eslint global-require: 0 */
'use strict';
const path = require('path');
const messages = require(path.join(__dirname, '..', '..', 'lib', 'messages'))();
module.exports = function (_execCommand) {
    return {
        description: messages.getMessage('commandDescription', [], 'projectUpgrade'),
        longDescription: messages.getMessage('commandDescriptionLong', [], 'projectUpgrade'),
        help: messages.getMessage('commandHelp', [], 'projectUpgrade'),
        requiresProject: false,
        flags: [{
                name: 'forceupgrade',
                char: 'f',
                description: messages.getMessage('forceOption', [], 'projectUpgrade'),
                longDescription: messages.getMessage('forceOptionLong', [], 'projectUpgrade'),
                required: false,
                type: 'flag'
            }],
        run(context) {
            const heroku = require('heroku-cli-util');
            const upgrades = require(path.join(__dirname, '..', '..', 'lib', 'project', 'upgrades'));
            return _execCommand({
                execute: () => upgrades(heroku.prompt, context.flags.forceupgrade)
            }, context);
        }
    };
};
