// salesforce-alm/dist/commands/auth/auth_web_login.js
/*
 * Copyright (c) 2016, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root  or https://opensource.org/licenses/BSD-3-Clause
 */
/* eslint global-require: 0 */
'use strict';
// Node
const path = require('path');
const util = require('util');
// Local
const messages = require(path.join(__dirname, '..', '..', 'lib', 'messages'))();
module.exports = function (_execCommand) {
    return {
        description: messages.getMessage('description', [], 'auth_weblogin'),
        longDescription: messages.getMessage('longDescription', [], 'auth_weblogin'),
        help: messages.getMessage('help', [], 'auth_weblogin'),
        requiresProject: false,
        flags: [{
                name: 'clientid',
                char: 'i',
                description: messages.getMessage('clientId', [], 'auth'),
                longDescription: messages.getMessage('clientIdLong', [], 'auth'),
                hasValue: true,
                type: 'string'
            },
            {
                name: 'instanceurl',
                char: 'r',
                description: messages.getMessage('instanceUrl', [], 'auth'),
                longDescription: messages.getMessage('instanceUrlLong', [], 'auth'),
                hasValue: true,
                type: 'url'
            },
            {
                name: 'setdefaultdevhubusername',
                char: 'd',
                description: messages.getMessage('setDefaultDevHub', [], 'auth'),
                longDescription: messages.getMessage('setDefaultDevHubLong', [], 'auth'),
                hasValue: false,
                type: 'flag'
            },
            {
                name: 'setdefaultusername',
                char: 's',
                description: messages.getMessage('setDefaultUsername', [], 'auth'),
                longDescription: messages.getMessage('setDefaultUsernameLong', [], 'auth'),
                hasValue: false,
                type: 'flag'
            },
            {
                name: 'setalias',
                char: 'a',
                description: messages.getMessage('setAlias', [], 'auth'),
                longDescription: messages.getMessage('setAliasLong', [], 'auth'),
                hasValue: true,
                type: 'string'
            },
            {
                name: 'disablemasking',
                description: messages.getMessage('disableMasking', [], 'auth_weblogin'),
                longDescription: messages.getMessage('disableMaskingLong', [], 'auth_weblogin'),
                hasValue: false,
                hidden: true,
                type: 'flag'
            }
        ],
        /**
         * the heroku cli hook
         * @param context - the cli context
         * @returns {*  }
         */
        run(context) {
            const heroku = require('heroku-cli-util');
            const AuthCommand = require(path.join(__dirname, '..', '..', 'lib', 'authCommand'));
            const authCommand = new AuthCommand();
            if (!util.isNullOrUndefined(context.flags.clientid)) {
                return heroku.prompt(messages.getMessage('stdin', [], 'auth_weblogin'), { mask: !context.flags.disablemasking })
                    .then((secret) => {
                    const map = new Map();
                    map.set('secret', secret);
                    return _execCommand(authCommand, context, map);
                });
            }
            else {
                return _execCommand(authCommand, context);
            }
        }
    };
};
