// salesforce-alm/dist/commands/auth/auth_sfdxurl_store.js
/*
 * Copyright (c) 2016, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root  or https://opensource.org/licenses/BSD-3-Clause
 */
/* eslint global-require: 0 */
'use strict';
// Node
const path = require('path');
// Local
const messages = require(path.join(__dirname, '..', '..', 'lib', 'messages'))();
const consts = require(path.join('..', '..', 'lib', 'constants'));
module.exports = function (_execCommand) {
    return {
        description: messages.getMessage('description', [], 'auth_sfdxurl'),
        longDescription: messages.getMessage('longDescription', [], 'auth_sfdxurl'),
        help: messages.getMessage('help', [consts.AUTH_URL_FORMAT1, consts.AUTH_URL_FORMAT2], 'auth_sfdxurl'),
        requiresProject: false,
        flags: [{
                name: 'sfdxurlfile',
                char: 'f',
                description: messages.getMessage('file', [], 'auth_sfdxurl'),
                longDescription: messages.getMessage('fileLong', [], 'auth_sfdxurl'),
                hasValue: true,
                required: true,
                type: 'filepath'
            },
            {
                name: 'setdefaultdevhubusername',
                char: 'd',
                description: messages.getMessage('setDefaultDevHub', [], 'auth'),
                longDescription: messages.getMessage('setDefaultDevHubLong', [], 'auth'),
                hasValue: false,
                type: 'flag'
            },
            {
                name: 'setdefaultusername',
                char: 's',
                description: messages.getMessage('setDefaultUsername', [], 'auth'),
                longDescription: messages.getMessage('setDefaultUsernameLong', [], 'auth'),
                hasValue: false,
                type: 'flag'
            },
            {
                name: 'setalias',
                char: 'a',
                description: messages.getMessage('setAlias', [], 'auth'),
                longDescription: messages.getMessage('setAliasLong', [], 'auth'),
                hasValue: true,
                type: 'string'
            }
        ],
        /**
         * the heroku cli hook
         * @param context - the cli context
         * @returns {*  }
         */
        run(context) {
            const AuthCommand = require(path.join(__dirname, '..', '..', 'lib', 'authCommand'));
            const authCommand = new AuthCommand();
            return _execCommand(authCommand, context);
        }
    };
};
