// salesforce-alm/dist/commands/source/source_open.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
const Messages = require("../../lib/force-cli-messages");
module.exports = function (_execCommand) {
    return {
        command: 'source:open',
        description: Messages.get('SourceOpenDescription'),
        longDescription: Messages.get('SourceOpenLongDescription'),
        help: Messages.get('SourceOpenHelp'),
        requiresProject: true,
        args: [],
        flags: [{
                name: 'sourcefile',
                char: 'f',
                hasValue: true,
                required: true,
                hidden: false,
                type: 'file',
                description: Messages.get('SourceOpenFileDescription'),
                longDescription: Messages.get('SourceOpenFileLongDescription')
            }, {
                name: 'urlonly',
                char: 'r',
                hasValue: false,
                required: false,
                hidden: false,
                type: 'flag',
                description: Messages.get('SourceOpenPathDescription'),
                longDescription: Messages.get('SourceOpenPathLongDescription')
            }],
        supportsTargetUsername: true,
        run(context) {
            return __awaiter(this, void 0, void 0, function* () {
                const SourceOpenCommand = require('../../lib/source/sourceOpenCommand').SourceOpenCommand;
                return yield _execCommand(new SourceOpenCommand(), context);
            });
        }
    };
};
