// salesforce-alm/dist/commands/source/source_convert.js
/*
 * Copyright, 1999-2017, salesforce.com
 * All Rights Reserved
 * Company Confidential
 */
/* eslint global-require: 0 */
'use strict';
// Node
const path = require('path');
// Local
const messages = require(path.join(__dirname, '..', '..', 'lib', 'messages'))();
module.exports = function (_execCommand) {
    return {
        description: messages.getMessage('description', null, 'source_convert'),
        longDescription: messages.getMessage('longDescription', null, 'source_convert'),
        help: messages.getMessage('help', null, 'source_convert'),
        initializeMetadataRegistry: true,
        requiresUsername: false,
        showProgress: true,
        flags: [
            {
                name: 'rootdir',
                char: 'r',
                description: messages.getMessage('rootParam', null, 'source_convert'),
                longDescription: messages.getMessage('rootParamLongDescription', null, 'source_convert'),
                required: false,
                hasValue: true,
                type: 'directory'
            },
            {
                name: 'outputdir',
                char: 'd',
                description: messages.getMessage('outputDirectoryParam', null, 'source_convert'),
                longDescription: messages.getMessage('outputDirectoryParamLongDescription', null, 'source_convert'),
                required: false,
                hasValue: true,
                type: 'directory'
            },
            {
                name: 'packagename',
                char: 'n',
                description: messages.getMessage('packageNameParam', null, 'source_convert'),
                longDescription: messages.getMessage('packageNameParamLongDescription', null, 'source_convert'),
                required: false,
                hasValue: true,
                type: 'string'
            }
        ],
        /**
         * the heroku cli hook
         * @param context - the cli context
         * @returns {* }
         */
        run(context) {
            const SourceConvertCommand = require(path.join(__dirname, '..', '..', 'lib', 'sourceConvertCommand'));
            return _execCommand(new SourceConvertCommand(), context);
        }
    };
};
