// salesforce-alm/dist/commands/org/org_open.js
/*
 * Copyright, 1999-2016, salesforce.com
 * All Rights Reserved
 * Company Confidential
 */
/* eslint global-require: 0 */
'use strict';
const path = require('path');
const messages = require(path.join(__dirname, '..', '..', 'lib', 'messages'))();
module.exports = function (_execCommand) {
    return {
        description: messages.getMessage('openCommandCliDescription'),
        longDescription: messages.getMessage('openCommandCliDescriptionLong'),
        help: messages.getMessage('openCommandCliHelp'),
        requiresProject: false,
        supportsTargetUsername: true,
        flags: [{
                name: 'path',
                char: 'p',
                description: messages.getMessage('openCommandCliPath'),
                longDescription: messages.getMessage('openCommandCliPathLong'),
                hasValue: true,
                required: false,
                type: 'string'
            },
            {
                name: 'urlonly',
                char: 'r',
                description: messages.getMessage('openCommandUrlonly'),
                longDescription: messages.getMessage('openCommandUrlonlyLong'),
                hasValue: false,
                required: false,
                type: 'flag'
            }
        ],
        /**
         * the heroku cli hook
         * @param context - the cli context
         * @returns {Promise}
         */
        run(context) {
            const OpenCommand = require(path.join(__dirname, '..', '..', 'lib', 'orgOpenCommand'));
            return _execCommand(new OpenCommand(), context);
        }
    };
};
