// salesforce-alm/dist/commands/org/org_shape_list.js
/*
 * Copyright (c) 2016, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root  or https://opensource.org/licenses/BSD-3-Clause
 */
/* eslint global-require: 0 */
'use strict';
const path = require('path');
const fs = require('fs');
const messages = require(path.join(__dirname, '..', '..', 'lib', 'messages'))();
module.exports = function (_execCommand) {
    return {
        description: messages.getMessage('description', [], 'org_shape_list'),
        longDescription: messages.getMessage('longDescription', [], 'org_shape_list'),
        help: messages.getMessage('help', [], 'org_shape_list'),
        requiresProject: false,
        flags: [{
                name: 'verbose',
                description: messages.getMessage('verbose', [], 'org_shape_list'),
                longDescription: messages.getMessage('verboseLong', [], 'org_shape_list'),
                hasValue: false,
                required: false
            }],
        /**
         * the heroku cli hook
         * @param context - the cli context
         * @returns {Promise}
         */
        run(context) {
            const Promise = require('bluebird');
            Promise.promisifyAll(fs);
            const OrgShapeListCommand = require(path.join(__dirname, '..', '..', 'lib', 'orgShapeListCommand')); // eslint-disable-line global-require
            return _execCommand(new OrgShapeListCommand(), context);
        }
    };
};
