// salesforce-alm/dist/commands/org/org_list.js
/*
 * Copyright (c) 2016, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root  or https://opensource.org/licenses/BSD-3-Clause
 */
/* eslint global-require: 0 */
'use strict';
const path = require('path');
const fs = require('fs');
const messages = require(path.join(__dirname, '..', '..', 'lib', 'messages'))();
module.exports = function (_execCommand) {
    return {
        description: messages.getMessage('description', [], 'org_list'),
        longDescription: messages.getMessage('longDescription', [], 'org_list'),
        help: messages.getMessage('help', [], 'org_list'),
        requiresProject: false,
        flags: [{
                name: 'verbose',
                description: messages.getMessage('verbose', [], 'org_list'),
                longDescription: messages.getMessage('verboseLong', [], 'org_list'),
                hasValue: false,
                required: false
            }, {
                name: 'all',
                description: messages.getMessage('all', [], 'org_list'),
                longDescription: messages.getMessage('allLong', [], 'org_list'),
                hasValue: false,
                required: false,
                type: 'flag'
            }, {
                name: 'clean',
                description: messages.getMessage('clean', [], 'org_list'),
                longDescription: messages.getMessage('cleanLong', [], 'org_list'),
                hasValue: false,
                required: false,
                type: 'flag'
            }, {
                name: 'noprompt',
                char: 'p',
                description: messages.getMessage('noPrompt', [], 'org_list'),
                longDescription: messages.getMessage('noPromptLong', [], 'org_list'),
                hasValue: false,
                required: false,
                type: 'flag'
            }],
        /**
         * the heroku cli hook
         * @param context - the cli context
         * @returns {Promise}
         */
        run(context) {
            const heroku = require('heroku-cli-util');
            const Promise = require('bluebird');
            const OrgListCommand = require(path.join(__dirname, '..', '..', 'lib', 'orgListCommand'));
            Promise.promisifyAll(fs);
            return _execCommand(new OrgListCommand(heroku.prompt), context);
        }
    };
};
