// salesforce-alm/dist/commands/org/org_delete.js
/*
 * Copyright (c) 2016, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root  or https://opensource.org/licenses/BSD-3-Clause
 */
/* eslint global-require: 0 */
'use strict';
const path = require('path');
const messages = require(path.join(__dirname, '..', '..', 'lib', 'messages'))();
module.exports = function (_execCommand) {
    return {
        description: messages.getMessage('deleteOrgCommandDescription'),
        longDescription: messages.getMessage('deleteOrgCommandDescriptionLong'),
        help: messages.getMessage('deleteOrgCommandHelp'),
        requiresProject: false,
        supportsTargetUsername: true,
        supportsTargetDevHubUsername: true,
        flags: [
            // override default 'targetusername' attributes (name enables param object merge)
            {
                name: messages.targetusername,
                description: messages.getMessage('usernameOption'),
                longDescription: messages.getMessage('usernameOptionLong'),
                required: true
            },
            {
                name: 'noprompt',
                char: 'p',
                description: messages.getMessage('deleteOrgCommandForceDescription'),
                longDescription: messages.getMessage('deleteOrgCommandForceDescriptionLong'),
                hasValue: false,
                required: false,
                type: 'flag'
            }
        ],
        /**
         * the heroku cli hook
         * @param context - the cli context
         * @returns {*  }
         */
        run(context) {
            const heroku = require('heroku-cli-util');
            const ActiveScratchOrgDeleteCommand = require(path.join(__dirname, '..', '..', 'lib', 'activeScratchOrgDeleteCommand'));
            const activeScratchOrgDeleteCommand = new ActiveScratchOrgDeleteCommand();
            // don't prompt yes/no if we're forcing the delete request, just execute
            if (context.flags.noprompt) {
                return _execCommand(activeScratchOrgDeleteCommand, context);
            }
            else {
                return heroku.prompt(messages.getMessage('deleteOrgCommandYesNo', context.flags.targetusername), {})
                    .then((answer) => {
                    if (answer.toUpperCase() === 'YES' || answer.toUpperCase() === 'Y') {
                        return _execCommand(activeScratchOrgDeleteCommand, context);
                    }
                    return undefined;
                });
            }
        }
    };
};
