// salesforce-alm/dist/commands/org/org_shape_create.js
/*
 * Copyright (c) 2017, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root  or https://opensource.org/licenses/BSD-3-Clause
 */
/* eslint global-require: 0 */
"use strict";
// Node
const path = require('path');
// Local
const messages = require(path.join(__dirname, '..', '..', 'lib', 'messages'))();
module.exports = function (_execCommand) {
    return {
        description: messages.getMessage('create_shape_command_description', null, 'org_shape'),
        longDescription: messages.getMessage('create_shape_command_description_long', null, 'org_shape'),
        help: messages.getMessage('create_shape_command_help', null, 'org_shape'),
        requiresProject: false,
        supportsTargetUsername: true,
        /**
         * the heroku cli hook
         * @param context - the cli context
         * @returns {*  }
         */
        run(context) {
            const ShapeRepCreateCommand = require(path.join(__dirname, '..', '..', 'lib', 'shapeRepCreateCommand'));
            const createCommand = new ShapeRepCreateCommand();
            return _execCommand(createCommand, context);
        }
    };
};
