// salesforce-alm/dist/commands/org/org_display.js
/*
 * Copyright (c) 2016, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root  or https://opensource.org/licenses/BSD-3-Clause
 */
/* eslint global-require: 0 */
'use strict';
const path = require('path');
const messages = require(path.join(__dirname, '..', '..', 'lib', 'messages'))();
module.exports = function (_execCommand) {
    return {
        description: messages.getMessage('displayCommandCliDescription'),
        longDescription: messages.getMessage('displayCommandCliDescriptionLong'),
        help: messages.getMessage('displayCommandCliHelp'),
        requiresProject: false,
        supportsTargetUsername: true,
        flags: [
            { name: 'verbose', hidden: false }
        ],
        /**
         * the heroku cli hook
         * @param context - the cli context
         * @returns {Promise}
         */
        run(context) {
            const OrgDisplayCommand = require(path.join(__dirname, '..', '..', 'lib', 'orgDisplayCommand'));
            return _execCommand(new OrgDisplayCommand(context), context);
        }
    };
};
