// salesforce-alm/dist/lib/messages.js
/*
 * Copyright (c) 2016, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root  or https://opensource.org/licenses/BSD-3-Clause
 */
"use strict";
const os = require('os');
const util = require('util');
const TARGET_USERNAME_PARAM = 'targetusername';
const messages = {
    default: {
        en_US: {
            // errors
            GeneralCommandFailure: 'This command failed.',
            InvalidProjectWorkspace: 'This directory does not contain a valid Salesforce DX project.',
            MissingClientConfig: 'Missing a client configuration. Please run the config command.',
            TestMessage: 'This is a test message do not change.',
            MissingAppConfig: 'The current project is invalid. sfdx-project.json is missing.',
            OldSfdxWorkspaceJsonPresent: 'Project file %s/sfdx-workspace.json needs to be renamed to sfdx-project.json',
            UndefinedLocalizationLabel: 'Missing label %s:%s for locale %s.',
            LoginServerNotFound: 'The hostname for the login server is not found: %s:%s',
            InvalidProjectDescriptor: 'This project descriptor is in invalid. The attribute [%s] is not found.',
            MissingRequiredParameter: 'A required parameter [%s] is missing.',
            InvalidParameter: 'Invalid [%s] parameter: %s',
            MissingScratchOrgNamespace: 'The [NamespacePrefix] is missing from the ScratchOrgInfo.',
            MaxOrgIds: 'The maximum number of local orgs (%s) has been reached',
            MaxOrgIdsAction: 'Use "sfdx force:org:list --clean" to remove outdated org configurations.',
            UserSessionIsInvalid: '%s: The user session for this org is expired or invalid. Please run config.',
            MissingScratchOrgDef: 'A scratch org definition file not found at %s. Please create it.',
            NonScratchOrgPush: 'You can push source only to scratch orgs using this command. Use Metadata API to sync changes for non-scratch orgs.',
            ProblemQueryingOrganizationSettingsDetail: 'No organization settings found',
            ProblemSettingOrgPrefs: 'Cannot set org preferences. %s',
            PathDoesNotExist: 'The specified path [%s] does not exist',
            InvalidArgumentFilePath: 'Invalid argument to --%s. You specified a directory path [%s], but a file is required.',
            InvalidArgumentDirectoryPath: 'Invalid argument to --%s. You specified a file path [%s], but a directory is required.',
            InvalidValueForDefaultPath: 'In sfdx-project.json, set the default to true or false. Example: [{ "path": "packageDirectory1", "default": true }, { "path": "packageDirectory2", "default": false }]',
            MultipleDefaultPaths: 'In sfdx-project.json, indicate only one package directory (path) as the default.',
            MissingDefaultPath: 'In sfdx-project.json, be sure to specify which package directory (path) is the default. Example: [{ "path": "packageDirectory1", "default": true }, { "path": "packageDirectory2" }]',
            InvalidPackageDirectory: 'The path %s, specified in sfdx-project.json, does not exist. Be sure this directory is included in your project root.',
            InvalidAbsolutePath: 'The path %s, specified in sfdx-project.json, must be indicated as a relative path to the project root.',
            IncorrectLogLevel: 'valid values are {%s}',
            LoggerNameRequired: 'A logger name is required',
            InvalidVariableReference: 'Invalid variable reference: variable "%s" not found or unset.',
            ValidationSchemaFieldErrors: 'Schema validation failed with following errors:\n%s',
            ValidationSchemaUnknown: 'Schema validation failed with unknown error',
            InvalidJson: 'An error occurred parsing "%s"',
            JsonParseError: 'Parse error in file %s on line %s\n%s\n',
            InvalidJsonCasing: 'All JSON input must have heads down camelcase keys.  E.g., { sfdcLoginUrl: "https://login.salesforce.com" }\nFound "%s" in\n%s',
            NoWorkspaceOrUser: `Command must be executed in a project directory or with the --${TARGET_USERNAME_PARAM} flag.`,
            PortInUse: 'Cannot start the OAuth redirect server on port %s.',
            PortInUseAction: 'Kill the process running on port %s or use a custom connected app and update OauthLocalPort in the sfdx-project.json file.',
            OrgDataNotAvailableError: 'An attempt to refresh the authentication token failed with a "Data Not Found Error". The org identified by username %s doesn\'t appear to exist. Likely cause is that the org was deleted by another user or has expired.',
            OrgDataNotAvailableErrorAction: 'Run "sfdx force:org:list --clean" to remove stale org authentications.\nUse "sfdx force:config" to update the defaultusername.\nUse "sfdx force:org:create" to create a new org.\nUse "sfdx force:auth" to authenticate an existing org.',
            InvalidCommandGroup: 'You have specified an invalid command group in which to stash your values. Please verify that you are specifying a valid command or register your command in stash.js under Stash#Commands.',
            // help
            schemaInfoOption: 'display schema information for the --%s configuration file to stdout; if you use this option, all other options except --json are ignored',
            schemaInfoOptionLong: 'Displays the schema information for the configuration file. If you use this option, all other options, except --json, are ignored.',
            invalidInstanceUrlForAccessTokenAction: 'Verify that the instanceUrl config setting is set to the instance that the access token belongs to.' +
                '\nUse "sfdx force:config:list" to view your current setting.' +
                '\nUse "sfdx force:config:set instanceUrl=<instance URL> --global" to set your instanceUrl to the correct instance.',
            cliForceConfigHelp: 'configures this project with a scratch org',
            cliForceConfigHelpClientId: 'specify the connected app client ID for the master org',
            cliForceConfigHelpClientSecret: 'specify the connected app client secret for the master org',
            cliForceConfigHelpClientRedirectUri: 'specify the connected app redirect uri for the master org',
            cliForceConfigHelpUsername: 'the master org username',
            cliForceConfigHelpPassword: 'the master org password',
            cliForceConfigHelpSuccess: 'successfully updated the SFDC accessToken',
            cliForceSyncHelp: 'synchronize Salesforce source between the scratch org and the project',
            cliForceSyncUpHelp: 'synchronize project source to the scratch org',
            cliForceSyncDownHelp: 'synchronize scratch org source to the project',
            cliForceHelp: 'tools for the Salesforce developer',
            cliForceSyncTypeHelp: '[All | Changed] Sync all or only the changed source. All is default.',
            cliForceRefreshHelp: 'Refreshes the auth token for the scratch org.',
            cliForceRefreshSuccess: 'Successfully reset the org auth token. Chive on!',
            cliForceCreateHelpType: 'The type of source to create. Values: [ScratchOrg]]',
            cliForceCreateHelpFile: 'Path to a file containing org signup parameters, in JSON format.',
            cliForceCreateHelpJson: 'The org signup parameters in JSON format.',
            cliForceCreateHelpPassword: 'Password for the admin user.',
            cliForceCreateMessageWrongType: 'Unsupported type. Valid types are [%s].',
            cliForceCreateMessageSuccess: 'Scratch org successfully created with id: %s',
            cliForceCreateMessagePassword: 'Setting org password...',
            cliForceCreateMessagePasswordOk: 'Ok',
            cliForceCreateNoConfig: 'Please specify an org configuration via file and/or key=value pairs',
            cliForceTestRunHelp: 'Invoke Apex tests in a given org.',
            cliForceTestRunHelp_results: 'Wait for results. Display results when tests are complete. If false, use "getTestResults -i [runTestId]" command to get results.',
            cliForceTestRunHelp_classIds: 'Apex Class Ids to test.',
            cliForceTestRunHelp_suiteIds: 'Test Suite Ids to test (WIP).',
            cliForceTestGetResultsHelp: 'Retrieve test results for given runTestId in a given org.',
            cliForceTestGetResultsHelp_id: 'Id of test run.',
            oauthBrowserSuccess: 'Successfully updated the auth configuration for the org.',
            closeTheBrowser: 'You may now close the browser.',
            urlStateMismatch: 'The Oauth redirect listener encountered an http request that is not trusted. Ignoring.',
            keyChainItemCreateFailed: 'Attempting to create an encryption key failed. %s',
            UnsupportedOperatingSystem: 'Unsupported Operating System: %s',
            invalidEncryptedFormat: 'The encrypted data is not properly formatted.',
            invalidEncryptedFormatAction: 'If attempting to create a scratch org then re-authorize. Otherwise create a new scratch org.',
            authDecryptFailed: 'Failed to decipher auth data. reason: %s.',
            genericTimeoutMessage: 'Socket timeout occurred while listening for results.',
            genericTimeoutCommandWaitMessageAction: 'Use command "%s" to retry. You may consider increasing --wait parameter value to increase timeout.',
            genericTimeoutWaitMessageAction: 'You may consider increasing the --wait parameter value to increase timeout.',
            subscriberHandshakeTimeout: 'Subscriber handshake failed due to a socket timeout.',
            subscriberHandshakeTimeoutAction: 'Check your connection to force.com and try again',
            herokuTopicDescription: 'tools for the Salesforce developer',
            streamingWait: 'the streaming client socket timeout (in minutes)',
            streamingWaitLong: 'Sets the streaming client socket timeout, in minutes.' +
                'If the streaming client socket has no contact from the server for a number of minutes, the client exits. ' +
                'Specify a longer wait time if timeouts occur frequently.',
            createOrgCommandDuration: 'duration of the scratch org (in days) (default:7, min:1, max:30)',
            createOrgCommandDurationLong: 'Sets the duration of the scratch org, in days. ' +
                'Valid values are from 1-30. The default is 7 days.',
            unrecognizedScratchOrgOption: '%s is not a supported option for scratch org configuration.',
            herokuNamespaceDescription: 'tools for the Salesforce developer',
            jsonOutputOption: 'format output as json',
            jsonOutputOptionLong: 'Format output as JSON.',
            loglevelOption: 'logging level for this command invocation',
            loglevelOptionLong: 'The logging level for this command invocation. Logs are stored in $HOME/.sfdx/sfdx.log.',
            usernameOption: 'username or alias for the target org',
            usernameOptionLong: 'A username or alias for the target org.',
            targetUsernameOption: 'username or alias for the target org; overrides default target org',
            targetUsernameOptionLong: 'A username or alias for the target org. Overrides the default target org.',
            invalidPortNumber: 'Invalid OAuth redirect port number defined: %s',
            ClientSecretRequired: 'The client secret is required.',
            authorizeCommandMissingJwtOption: 'Both username and file must be provided.',
            authorizeCommandMissingClientId: 'The client ID is required for the JWT-based authentication process',
            authorizeCommandSuccess: 'Successfully authorized %s with org ID %s',
            authorizeCommandCloseBrowser: 'You may now close the browser',
            createOrgCommandSuccess: 'Successfully created scratch org: %s, username: %s',
            createOrgCommandDescription: 'create a scratch org',
            createOrgCommandDescriptionLong: 'Creates a scratch org using values specified in a configuration file or ' +
                'key=value pairs that you specify on the command line. Values specified on the command line override values in the configuration file.',
            createOrgCommandHelp: 'To set up a connected app for your new scratch org, specify the value that was ' +
                'returned when you created a connected app in your Dev Hub org as --clientid.' +
                '\n\nExamples:' +
                '\n   $ sfdx force:org:create -f config/enterprise-scratch-def.json -a TestOrg1' +
                '\n   $ sfdx force:org:create -a MyDevOrg -s -v me@myhub.org edition=Developer' +
                '\n   $ sfdx force:org:create -f config/enterprise-scratch-def.json -a OrgWithOverrides username=testuser1@mycompany.org',
            createOrgCommandClientId: 'connected app consumer key',
            createOrgCommandClientIdLong: 'Connected app consumer key, as configured in your Dev Hub.',
            createOrgCommandSet: 'set the created org as the default username',
            createOrgCommandSetLong: 'Sets the created org as the default username.',
            createOrgCommandAlias: 'set an alias for for the created scratch org',
            createOrgCommandAliasLong: 'Sets an alias for for the created scratch org.',
            createOrgTargetDevhubUsername: 'username or alias for the dev hub org; overrides default dev hub org',
            createOrgTargetDevhubUsernameLong: 'A username or alias for the target Dev Hub org. Overrides the default Dev Hub org.',
            createOrgCommandFile: 'path to a scratch org definition file',
            createOrgCommandFileLong: 'Path to a scratch org definition file. Either --definitionfile or ' +
                '--definitionjson is required.',
            createOrgCommandObject: 'scratch org definition in json format ',
            createOrgCommandObjectLong: 'Scratch org definition in JSON format. Either --definitionfile or ' +
                '--definitionjson is required.',
            createOrgCommandNoNamespace: 'creates the scratch org with no namespace',
            createOrgCommandNoNamespaceLong: 'Creates the scratch org with no namespace. Useful when using a scratch ' +
                'org to test installations of packages with namespaces.',
            createOrgCommandNoAncestors: 'do not include second-generation package ancestors in the scratch org',
            createOrgCommandNoAncestorsLong: 'Do not include second-generation package ancestors in the scratch org.',
            createOrgCommandEnv: 'environment where the scratch org is created: [%s]',
            createOrgCommandEnvLong: 'Environment where the scratch org is created: [%s].',
            createOrgCommandUnauthorized: 'You do not have access to the [%s] object',
            createOrgTimeoutHintNoIdAction: 'Retry creating the org but increase the wait timeout.',
            apexReportCommandNoTestFound: 'No test found with the given names [%s]',
            apexReportCommandNoJobId: 'No test run job ID found.',
            apexReportCommandInvalidJobId: 'Invalid test run job ID: ',
            apexReportCommandInvalidResponse: 'Invalid API response: %s',
            apexReportCommandTestResultRetrievalFailed: 'Unable to get test results for job [%s]: [%s]',
            apexReportCommandCodeCoverageRetrievalFailed: 'Unable to get code coverage for job [%s]: [%s]',
            CoverageWithoutReporter: 'A result format is required when specifying codecoverage',
            apexTestApiInvalidReporter: 'Unknown result format type. Must be one of the following values: %s',
            apexTestApiInvalidTestRunId: 'Test run job ID not found: %s',
            apexTestApiInvalidTestLevel: 'Unknown testlevel `%s`. Must be one of the following values: [%s]',
            apexTestApiIncorrectTestLevel: 'When specifying classnames or suitenames the provided testlevel must be RunSpecifiedTests',
            apexTestApiInvalidParams: 'Can not specify both classnames and suitenames',
            InvalidAsyncTestJob: 'Unable to invoke async test job: %s',
            InvalidAsyncTestJobUnknownAction: 'Try running the tests in the Developer Console by using force:org:open, or create a new scratch org and try again.',
            InvalidAsyncTestJobNoneFound: 'Unable to invoke any tests. Ensure the tests are loaded into the org or valid inputs are supplied.',
            unsupportedValueForEnv: 'Unsupported value for env: [%s]',
            unsupportedValueForDuration: 'Unsupported value for durationdays (must be 1-30): [%s]',
            invalid_client: 'Invalid client credentials. Verify the OAuth client secret and ID.',
            signupFailed: 'The request to create a scratch org failed with error code: %s.',
            signupUnexpected: 'The request to create a scratch org returned an unexpected status',
            signupFieldsMissing: 'Required fields are missing for org creation: [%s]',
            pushCommandAsyncRequestInvalidated: 'Salesforce cancelled the job because the results might not be valid. Is there a newer compile request?',
            pushCommandAsyncRequestUnexpected: 'An unexpected error occurred during deploy.',
            pushCommandCliDescription: 'push source to an org from the project',
            pushCommandCliDescriptionLong: 'Pushes changed source from your project to the scratch org to keep them in sync.',
            pushCommandCliHelp: 'If the command detects a conflict, it displays the conflicts but does not complete the process. ' +
                'After reviewing the conflict, rerun the command with the --forceoverwrite parameter.',
            pushCommandCliInvalidUsernameOption: 'Invalid value for username',
            pushCommandCliPreExecute: 'Pushing source changes to org %s as user %s',
            pushCommandWaitOptionDescriptionLong: 'Number of minutes to wait for the command to complete and display ' +
                'results to the terminal window. If the command continues to run after the wait period, the CLI ' +
                'returns control of the terminal window to you. The default is 33 minutes.',
            pushCommandForceOption: 'ignore conflict warnings and overwrite changes to scratch org',
            pushCommandForceOptionLong: 'Runs the push command even if conflicts exist. Changes in the project overwrite changes in the scratch org.',
            pushCommandConflictMsg: 'We couldn\'t complete the push operation due to conflicts. Verify that you want to keep the local versions, then run "sfdx force:source:push -f" with the --forceoverwrite (-f) option.',
            pullCommandConflictMsg: 'We couldn\'t complete the pull operation due to conflicts. Verify that you want to keep the remote versions, then run "sfdx force:source:pull -f" with the --forceoverwrite (-f) option.',
            pushCommandHumanSuccess: 'Pushed Source',
            MissingMetadataExtension: 'Expected file at source path: %s to end with the \'-meta.xml\' extension. Please rename the file to %s',
            MissingMetadataFileWithMetaExt: 'Expected metadata file with \'-meta.xml\' extension at source path: %s',
            MissingMetadataFile: 'Expected metadata file at source path: %s',
            MissingContentFile: 'Expected content file(s) at source path(s): %s',
            UnsupportedMimeTypes: 'The following MIME types are not supported: %s',
            statusCommandCliDescription: 'list local changes and/or changes in a scratch org',
            statusCommandCliLongDescription: 'Lists changes that have been made locally, in a scratch org, or both.',
            statusCommandCliHelp: 'Examples:' +
                '\n   $ sfdx force:source:status -l' +
                '\n   $ sfdx force:source:status -r' +
                '\n   $ sfdx force:source:status -a' +
                '\n   $ sfdx force:source:status -a -u me@my.org --json',
            statusCommandAllOptionDescription: 'list all the changes that have been made',
            statusCommandAllOptionDescriptionLong: 'Lists all the changes that have been made.',
            statusCommandLocalOptionDescription: 'list the changes that have been made locally',
            statusCommandLocalOptionDescriptionLong: 'Lists the changes that have been made locally.',
            statusCommandRemoteOptionDescription: 'list the changes that have been made in the scratch org',
            statusCommandRemoteOptionDescriptionLong: 'Lists the changes that have been made in the scratch org.',
            statusCommandHumanSuccess: 'Source Status',
            mdapiPullCommandNoDataReturned: 'No metadata was returned by the retrieve',
            // mdapi general-purpose messages
            mdapiCliInvalidNumericParam: 'Invalid value for %s. Must be a positive numerical value.',
            mdapiCliWaitTimeExceededError: 'Your %s request did not complete within the specified wait time [%s minutes]. Try again with a longer wait time.',
            mdapiCliWait: 'wait time for command to finish in minutes (default: %s)',
            mdapiCliWaitLong: 'The number of minutes to wait for the command to complete. The default is –1 (no limit).',
            mdapiCliExclusiveFlagError: 'Specify either --%s or --%s but not both.',
            mdapiCliInvalidWaitError: 'Specify the number of minutes to wait as a numerical value greater than or equal to -1. You can specify a decimal value if it is greater than 0.',
            mdapiDeployFailed: 'The metadata deploy operation failed.',
            // mdapi main topic short and long descriptions
            mdapiTopicDescription: 'retrieve and deploy metadata using Metadata API',
            mdapiTopicLongDescription: 'Use the mdapi commands to retrieve and deploy Metadata API–formatted files that ' +
                'represent components in an org, or to convert Metadata API–formatted metadata into the source format used in ' +
                'Salesforce DX projects.',
            // mdapi:deploy:report short and long descriptions, help, and parameter short and long descriptions
            mdDeployReportCommandCliDescription: 'check the status of a metadata deployment',
            mdDeployReportCommandCliLong: 'Checks the current status of an asynchronous metadata deployment.',
            mdDeployReportCommandCliHelp: 'Specify the job ID for the deploy you want to check. You can also specify a wait time ' +
                '(minutes) to check for updates to the deploy status.',
            mdDeployReportCommandCliVerboseLong: 'Indicates that you want verbose output for deploy results.',
            // mdapi:deploy short and long descriptions, help, and parameter short and long descriptions
            mdDeployCommandCliDescription: 'deploy metadata to an org using Metadata API',
            mdDeployCommandCliLong: 'Deploys file representations of components into an org by creating or updating the ' +
                'components they represent. You can deploy and retrieve up to 10,000 files or 400 MB (39 MB compressed) ' +
                'at one time. The default target username is the admin user for the default scratch org.',
            mdDeployCommandCliHelp: 'Specify the location of the files to deploy as a .zip file or by the root of the ' +
                'directory tree containing the files. To check the status of a deployment, specify its job ID.' +
                '\n\nThe default value of --rollbackonerror is true, but the corresponding parameter in the Metadata API ' +
                'deploy() call defaults to false.',
            mdDeployCommandCliDeployDir: 'root of directory tree of files to deploy',
            mdDeployCommandCliDeployDirLong: 'The root of the directory tree that contains the files to deploy. The root ' +
                'must contain a valid package.xml file describing the entities in the directory structure. Required to ' +
                'initiate a deployment if you don’t use --zipfile. If you specify both --zipfile and --deploydir, a zip ' +
                'file of the contents of the --deploydir directory is written to the location specified by --zipfile.',
            mdDeployCommandCliCheckOnly: 'validate deploy but don’t save to the org',
            mdDeployCommandCliCheckOnlyLong: 'Validates the deployed metadata and runs all Apex tests, but prevents the ' +
                'deployment from being saved to the org.' +
                '\nIf you change a field type from Master-Detail to Lookup or vice versa, that change isn’t supported when ' +
                'using the --checkonly parameter to test a deployment (validation). This kind of change isn’t supported for ' +
                'test deployments to avoid the risk of data loss or corruption. If a change that isn’t supported for test ' +
                'deployments is included in a deployment package, the test deployment fails and issues an error.' +
                '\nIf your deployment package changes a field type from Master-Detail to Lookup or vice versa, you can ' +
                'still validate the changes prior to deploying to Production by performing a full deployment to another ' +
                'test Sandbox. A full deployment includes a validation of the changes as part of the deployment process.',
            mdDeployCommandCliJobId: 'job ID of the deployment you want to check',
            mdDeployCommandCliJobIdLong: 'The job ID (asyncId) of the deployment you want to check. Use with -w to resume waiting.',
            mdDeployCommandCliTestLevel: 'deployment testing level',
            mdDeployCommandCliTestLevelLong: 'Specifies which level of deployment tests to run. Valid values are:\n' +
                'NoTestRun—No tests are run. This test level applies only to deployments to development environments, ' +
                'such as sandbox, Developer Edition, or trial orgs. This test level is the default for development ' +
                'environments.\n' +
                'RunSpecifiedTests—Runs only the tests that you specify in the --runtests option. Code coverage ' +
                'requirements differ from the default coverage requirements when using this test level. Executed tests ' +
                'must comprise a minimum of 75% code coverage for each class and trigger in the deployment package. This ' +
                'coverage is computed for each class and trigger individually and is different than the overall coverage ' +
                'percentage.\n' +
                'RunLocalTests—All tests in your org are run, except the ones that originate from installed managed ' +
                'packages. This test level is the default for production deployments that include Apex classes or triggers.\n' +
                'RunAllTestsInOrg—All tests in your org are run, including tests of managed packages.\n' +
                'If you don’t specify a test level, the default behavior depends on the contents of your deployment ' +
                'package. For more information, see “Running Tests in a Deployment” in the Metadata API Developer Guide.',
            mdDeployCommandCliJobIdDeprecated: 'Instead, use "sfdx force:mdapi:deploy:report -i <jobId>".',
            mdDeployCommandCliRunTests: 'tests to run if --testlevel RunSpecifiedTests',
            mdDeployCommandCliRunTestsLong: 'Lists the Apex classes containing the deployment tests to run. Use this parameter when you set --testlevel to RunSpecifiedTests.',
            mdDeployCommandCliRollbackOnError: 'roll back deployment for any failure',
            mdDeployCommandCliRollbackOnErrorLong: 'Indicates whether any failure causes a complete rollback of the ' +
                'deploy operation. The default is true. If set to false, the operation performs actions that don’t have ' +
                'errors and returns errors for the remaining actions. You must set this parameter to true if you are ' +
                'deploying to a production org.',
            mdDeployCommandCliRollbackOnErrorDeprecated: 'Instead, use "ignoreerrors".',
            mdDeployCommandCliIgnoreErrors: 'ignore any errors and do not roll back deployment',
            mdDeployCommandCliIgnoreErrorsLong: 'Ignores the deploy errors, and continues with the ' +
                'deploy operation. The default is false. Keep this parameter set to false when deploying to a production org. ' +
                'If set to true, components without errors are deployed, and components with errors are skipped.',
            mdDeployCommandCliIgnoreWarnings: 'whether a warning will allow a deployment to complete successfully',
            mdDeployCommandCliIgnoreWarningsLong: 'If a warning occurs and ignoreWarnings is set to true, the success field in ' +
                'DeployMessage is true. When ignoreWarnings is set to false, success is set to false, and the warning is treated like an error.\n' +
                'This field is available in API version 18.0 and later. Prior to version 18.0, there was no distinction between warnings and errors. ' +
                'All problems were treated as errors and prevented a successful deployment.',
            mdDeployCommandCliZipFile: 'path to .zip file of metadata to deploy',
            mdDeployCommandCliZipFileLong: 'The path to the .zip file of metadata files to deploy. You must indicate this option or --deploydir.' +
                'If you specify both --zipfile and --deploydir, a .zip file of the contents of the deploy directory is created at the path ' +
                'specified for the .zip file.',
            mdDeployCommandCliVerbose: 'verbose output of deploy results',
            mdDeployCommandCliVerboseLong: 'Indicates that you want verbose output from the deploy operation.',
            // mdapi:deploy messages
            mdDeployCommandCliInvalidUsernameOption: 'That target username doesn’t exist. Try again with a valid target username.',
            mdDeployCommandCliZipFileError: 'The --zipfile parameter requires a file path to a zip file. Try again.',
            mdDeployCommandCliWaitTimeExceededError: 'The deploy request did not complete within the specified wait time [%s minutes].\n' +
                'To check the status of this deployment, run "sfdx force:mdapi:deploy:report"',
            mdDeployCommandCliInvalidJobIdError: 'The job for [%s] doesn’t exist. Try again with a valid job ID.',
            mdDeployCommandCliInvalidRollbackError: 'The rollbackonerror parameter cannot be %s. Specify either "true" or "false"',
            // mdapi:retrieve:report short and long descriptions, help, and parameter short and long descriptions
            mdRetrieveReportCommandCliHelp: 'Specify the job ID and a target directory for the retrieve you want to check. You can also ' +
                'specify a wait time (minutes) to check for updates to the deploy status. If the retrieve was successful, ' +
                'the resulting zip file will be saved to the location passed in with the retrieve target parameter.',
            mdRetrieveReportCommandCliDescription: 'check the status of a metadata retrieval',
            mdRetrieveReportCommandCliLong: 'Check the status of an asynchronous metadata retrieval.',
            // mdapi:retrieve short and long descriptions, help, and parameter short and long descriptions
            mdRetrieveCommandCliDescription: 'retrieve metadata from an org using Metadata API',
            mdRetrieveCommandCliLong: 'Uses Metadata API to retrieve a .zip of XML files that represent metadata from ' +
                'the targeted org. The default target username is the admin user for the default scratch org. You can ' +
                'retrieve and deploy up to 10,000 files or 400 MB (39 MB compressed) at one time.',
            mdRetrieveCommandCliHelp: 'The default target username is the admin user for the default scratch org. You can ' +
                'retrieve and deploy up to 10,000 files or 400 MB (39 MB compressed) at one time.',
            mdRetrieveCommandCliRetrieveTarget: '(required) root of the directory structure into which the .zip or ' +
                'metadata files are retrieved',
            mdRetrieveCommandCliRetrieveTargetDir: 'directory root for the retrieved files',
            mdRetrieveCommandCliRetrieveTargetDirLong: 'The root of the directory structure where the retrieved .zip or ' +
                'metadata files are put.',
            mdRetrieveCommandCliUnpackaged: 'file path of manifest of components to retrieve',
            mdRetrieveCommandCliUnpackagedLong: 'The complete path for the manifest file that specifies the components ' +
                'to retrieve.',
            mdRetrieveCommandCliUnzip: 'unzip file of retrieved components to directory',
            mdRetrieveCommandCliUnzipLong: 'Indicates whether to unzip the file of retrieved components to the ' +
                'retrievetargetdir directory (true) or to save the zip file in the retrievetargetdir directory (false).',
            mdRetrieveCommandCliVerbose: 'verbose output of retrieve result',
            mdRetrieveCommandCliVerboseLong: 'Indicates that you want verbose output from the retrieve operation.',
            mdRetrieveCommandCliSourceDir: 'source dir to use instead of default manifest',
            mdRetrieveCommandCliSourceDirLong: 'The source directory to use instead of the default manifest specified in ' +
                'sfdx-config.json.',
            mdRetrieveCommandCliSinglePackage: 'a single-package retrieve (default: false)',
            mdRetrieveCommandCliSinglePackageLong: 'Specifies whether only a single package is being retrieved (true) ' +
                'or more than one package (false).',
            mdRetrieveCommandCliPackageNames: 'a comma-separated list of packages to retrieve',
            mdRetrieveCommandCliPackageNamesLong: 'A comma-separated list of package names to retrieve.',
            mdRetrieveCommandCliJobId: 'job ID of the retrieve you want to check',
            mdRetrieveCommandCliJobIdLong: 'The job ID (asyncId) of the retrieve you want to check. You must specify a ' +
                '--retrievetargetdir. Use with --wait to resume waiting.',
            mdRetrieveCommandCliJobIdDeprecated: 'Instead, use "sfdx force:mdapi:retrieve:report -i <jobId> -r <targetDir>".',
            mdRetrieveCommandCliApiVersion: 'target API version for the retrieve (default %s)',
            mdRetrieveCommandCliApiVersionLong: 'Use to override the default, which is the latest version supported by ' +
                'your CLI plug-in, with the version in your package.xml file.',
            mdRetrieveCommandCliWait: 'wait time for command to finish in minutes (default: %s (no limit))',
            mdRetrieveCommandCliWaitLong: 'The number of minutes to wait for the command to complete.',
            // mdapi:retrieve messages
            mdRetrieveCommandCliInvalidUsernameOption: 'That target username doesn’t exist. Try again with a valid target username.',
            mdRetrieveCommandCliWaitTimeExceededError: 'The retrieve request did not complete within the specified wait time [%s minutes].\n' +
                'To check the status of this retrieve, run "sfdx force:mdapi:retrieve:report"',
            mdRetrieveCommandCliTooManyPackagesError: 'You specified [%s]. Try again and specify only one package when using --singlepackage.',
            mdRetrieveCommandCliInvalidProjectError: 'You can’t create a manifest from an artifact when you’re not in a project. ' +
                'Move into a valid project or specify the manifest location using a parameter (--packagenames, --unpackaged).',
            mdRetrieveCommandCliInvalidApiVersionError: 'Specify the API version as a positive numerical value less than or equal to the current API version (%s).',
            mdapiRetrieveFailed: 'The metadata retrieve operation failed: %s',
            pollTimeout: 'polling timeout in milliseconds (default %s ms)',
            pollInterval: 'polling interval in milliseconds (default %s ms)',
            waitParamValidValueError: 'Invalid value was specified for wait. Please provide a wait value greater than %s minutes.',
            stateTableColumn: 'State',
            fullNameTableColumn: 'Full Name',
            typeTableColumn: 'Type',
            workspacePathTableColumn: 'Project Path',
            orgTopicDescription: 'manage your Salesforce DX orgs',
            orgTopicDescriptionLong: 'Use the org commands to manage the orgs you use with Salesforce DX. Create and delete scratch orgs, ' +
                'list your created and authorized orgs, and open orgs in your browser.',
            orgShapeTopicDescription: 'manage your sfdx org shapes',
            orgShapeTopicDescriptionLong: 'Use the orgshape commands to manage the org shapes you use with Salesforce DX. Create and delete org shapes, ' +
                'and list your created org shapes.',
            openCommandCliDescription: 'open an org in your browser',
            openCommandCliDescriptionLong: 'Opens an org in your browser.',
            openCommandCliHelp: 'Opens your default scratch org, or another org that you specify.' +
                '\n\nTo open a specific page, specify the portion of the URL after "yourInstance.salesforce.com/" as --path. ' +
                '\nFor example, specify "--path one/one.app" to open Lightning Experience, or specify "--path /apex/YourPage" to open ' +
                'a Visualforce page.' +
                '\n\nTo generate a URL but not launch your browser, specify --urlonly.' +
                '\n\nExamples:' +
                '\n  $ sfdx force:org:open' +
                '\n  $ sfdx force:org:open -u me@my.org' +
                '\n  $ sfdx force:org:open -u MyTestOrg1' +
                '\n  $ sfdx force:org:open -r -p one/one.app',
            openCommandCliPath: 'navigation URL path',
            openCommandCliPathLong: 'Navigation URL path (not including domain).',
            openCommandUrlonly: 'display navigation URL, but don’t launch browser',
            openCommandUrlonlyLong: 'Displays a navigation URL, but doesn’t launch your browser.',
            openCommandHumanSuccess: 'Access org %s as user %s with the following URL: %s',
            openCommandDomainWaiting: 'Waiting to resolve the Lightning Experience-enabled custom domain...',
            openCommandDomainTimeoutError: 'The Lightning Experience-enabled custom domain is unavailable.',
            openCommandDomainTimeoutAction: 'The Lightning Experience-enabled custom domain may take a few more minutes to resolve. Try the "force:org:open" command again.',
            displayCommandCliDescription: 'get org description',
            displayCommandCliDescriptionLong: 'Gets the description for the current or target org.',
            displayCommandCliHelp: 'Output includes your access token, client ID, connected status, ' +
                'org ID, instance URL, username, and alias, if applicable. ' +
                'Use --verbose to include the SFDX auth URL.' +
                '\n\nExamples:' +
                '\n   $ sfdx force:org:display' +
                '\n   $ sfdx force:org:display -u me@my.org' +
                '\n   $ sfdx force:org:display -u TestOrg1 --json' +
                '\n   $ sfdx force:org:display -u TestOrg1 --json > tmp/MyOrgDesc.json',
            displayCommandHumanSuccess: 'Org Description',
            userTopicDescription: 'perform user-related admin tasks',
            userTopicDescriptionLong: 'Use the user commands to perform user-related admin tasks.',
            assignCommandCliDescription: 'assign a permission set to one or more users of an org',
            assignCommandCliDescriptionLong: 'Assigns a named permission set to one or more users of an org.',
            assignCommandCliHelp: 'Defaults to the defaultusername.' +
                '\n\nExamples:' +
                '\n   $ sfdx force:user:permset:assign -n DreamHouse' +
                '\n   $ sfdx force:user:permset:assign -n DreamHouse -u me@my.org' +
                '\n   $ sfdx force:user:permset:assign -n DreamHouse -o user1@my.org,user2,user3',
            assignCommandCliName: 'the name of the permission set to assign',
            assignCommandCliNameLong: 'The name of the permission set to assign.',
            assignCommandPermissionSetNotFoundError: 'Permission Set "%s" not found in target org. Do you need to push source?',
            assignCommandOnBehalfOf: 'comma-separated list of usernames or aliases to assign the permission set to',
            assignCommandOnBehalfOfLong: 'Comma-separated list of usernames or aliases to assign the permission set to.',
            pullCommandCliDescription: 'pull source from the scratch org to the project',
            pullCommandCliDescriptionLong: 'Pulls changed source from the scratch org to your project to keep them in sync.',
            pullCommandCliHelp: 'If the command detects a conflict, it displays the conflicts but does not complete the process. ' +
                'After reviewing the conflict, rerun the command with the --forceoverwrite parameter.',
            pullCommandMetadataTypeLabel: 'Metadata Type',
            pullCommandMetadataTypePath: 'Metadata Path',
            pullCommandCliPreExecute: 'Pulling source changes from org %s as user %s',
            pullCommandWaitOptionDescriptionLong: 'The number of minutes to wait for the command to complete and display ' +
                'results to the terminal window. If the command continues to run after the wait period, the CLI returns ' +
                'control of the terminal window to you. The default is 33 minutes.',
            pullCommandForceOptionDescription: 'ignore conflict warnings and overwrite changes to the project',
            pullCommandForceOptionDescriptionLong: 'Runs the pull command even if conflicts exist. Changes in the scratch org overwrite changes in the project.',
            pullCommandHumanSuccess: 'Pulled Source',
            packageCliDescription: 'install and uninstall first- and second-generation packages',
            packageCliDescriptionLong: 'Use the package commands to install and uninstall first- and second-generation packages.',
            package1CliDescription: 'develop first-generation managed and unmanaged packages',
            package1CliDescriptionLong: 'Use the package1 commands to create and view first-generation package versions in your Dev Hub org.',
            package1VersionCreateCommandCliDescription: 'create a first-generation package version in the release org',
            package1VersionCreateCommandCliDescriptionLong: 'Creates a first-generation package version in the release org.',
            package1VersionCreateCommandCliHelp: 'The package version is based on the contents of the specified metadata package. Omit -m if you want to create an unmanaged package version.',
            package1VersionCreateCommandId: 'ID of the metadata package (starts with 033) of which you’re creating a new version',
            package1VersionCreateCommandIdLong: 'ID of the metadata package (starts with 033) of which you’re creating a new version.',
            package1VersionCreateCommandName: 'package version name',
            package1VersionCreateCommandNameLong: 'Package version name.',
            package1VersionCreateCommandDescription: 'package version description',
            package1VersionCreateCommandDescriptionLong: 'Package version description.',
            package1VersionCreateCommandVersion: 'package version in major.minor format, for example, 3.2',
            package1VersionCreateCommandVersionLong: 'Package version in major.minor format, for example, 3.2.',
            package1VersionCreateCommandReleaseNotes: 'release notes URL',
            package1VersionCreateCommandReleaseNotesLong: 'Release notes URL.',
            package1VersionCreateCommandPostInstall: 'post install URL',
            package1VersionCreateCommandPostInstallLong: 'Post install URL.',
            package1VersionCreateCommandManagedReleased: 'create a managed package version',
            package1VersionCreateCommandManagedReleasedLong: 'Creates a managed package version. To create a beta version, don’t include this parameter.',
            package1VersionCreateCommandInstallationKey: 'installation key for key-protected package (default: null)',
            package1VersionCreateCommandInstallationKeyLong: 'Installation key for creating the key-protected package. The default is null.',
            package1VersionCreateCommandWait: 'minutes to wait for the package version to be created (default: 2 minutes)',
            package1VersionCreateCommandWaitLong: 'Minutes to wait for the package version to be created. The default is 2 minutes.',
            package1VersionCreateCommandNotANumber: 'Field %s must contain only a numeric value: %s.',
            package1VersionCreateCommandInvalidVersion: 'Version supplied, %s, is not formatted correctly. Enter in major.minor format, for example, 3.2.',
            package1VersionCreateCommandTimeout: 'Stopped waiting for package upload to finish. Wait time exceeded. waitTimeInMinutes = %s.',
            package1VersionCreateCommandUploadFailure: 'Package upload failed.',
            package1VersionCreateHumanSuccess: 'Successfully created package version: %s for package %s.',
            package1VersionListCommandLongDescription: 'Lists the versions for the specified package or all first-generation packages in the org.',
            package1VersionListCommandCliDescription: 'list package versions for the specified first-generation package or for the org',
            package1VersionListCommandCliHelp: 'If a metadata package ID is specified, lists all versions of the specified package. ' +
                'Otherwise, lists all package versions for the org. For each package version, the list includes the package ' +
                'version ID, metadata package ID, name, version number, and release state.',
            package1VersionListCommandPackageId: 'metadata package ID (starts with 033)',
            package1VersionListCommandPackageIdLong: 'Metadata package ID (starts with 033) whose package versions you want to list. If not specified, shows all versions for all packages (managed and unmanaged) in the org.',
            package1VersionListHumanSuccess: 'Successfully created the package version list.',
            package1VersionListAction: 'Verify that you entered a valid package ID and that you are authorized in the org. Then try again.',
            package1VersionDisplayCommandLongDescription: 'Displays detailed information about an individual first-generation package version.',
            package1VersionDisplayCommandCliDescription: 'display details about a first-generation package version',
            package1VersionDisplayCommandCliHelp: 'Display detailed information about an individual package version, including metadata package ID, name, the release state, and build number.',
            package1VersionDisplayCommandPackageId: 'metadata package version ID (starts with 04t)',
            package1VersionDisplayCommandPackageIdLong: 'ID (starts with 04t) of the metadata package version whose details you want to display.',
            package1VersionDisplayHumanSuccess: 'Successfully displayed the package version.',
            package1VersionDisplayAction: 'Verify that you entered a valid package version ID and try again.',
            deleteOrgCommandDescription: 'mark a scratch org for deletion',
            deleteOrgCommandDescriptionLong: 'Marks a scratch org for deletion.',
            deleteOrgCommandHelp: 'To mark the org for deletion without being prompted to confirm, specify --noprompt.' +
                '\n\nExamples:' +
                '\n   $ sfdx force:org:delete -u me@my.org' +
                '\n   $ sfdx force:org:delete -u MyOrgAlias -p',
            attemptingToDeleteExpiredOrDeleted: 'Attempting to delete an expired or deleted org',
            deleteOrgConfigOnlyCommandSuccess: 'Successfully deleted scratch org %s.',
            deleteOrgCommandSuccess: 'Successfully marked scratch org %s for deletion',
            deleteOrgCommandQueryError: 'Error querying for DevHubMember %s. We received %s results',
            deleteOrgCommandPathError: 'The scratch org config for scratch org %s does not exist',
            deleteOrgCommandYesNo: 'Enqueue scratch org with name: %s for deletion?  Are you sure (y/n)?',
            deleteOrgCommandForceDescription: 'no prompt to confirm deletion',
            deleteOrgCommandForceDescriptionLong: 'No prompt to confirm deletion.',
            deleteOrgHubError: 'The Dev Hub org cannot be deleted.',
            defaultOrgNotFound: 'No %s org found',
            defaultOrgNotFoundAction: 'Run the "sfdx force:auth" commands with --setdefaultusername to connect to an org and set it ' +
                'as your default org.\nRun "force:org:create" with --setdefaultusername to create a scratch org and set it as ' +
                'your default org.\nRun "sfdx force:config:set defaultusername=<username>" to set your default username.',
            defaultOrgNotFoundDevHubAction: 'Run the "sfdx force:auth" commands with --setdefaultdevhubusername to connect to a ' +
                'Dev Hub org and set it as your default Dev Hub.\nRun "force:org:list" to see a list of locally registered orgs.' +
                '\nRun "sfdx force:config:set defaultdevhubusername=<username>" to set your default Dev Hub username.',
            namedOrgNotFound: 'No org configuration found for name %s',
            noResultsFound: 'No results found',
            invalidVersionString: 'Invalid API version string.',
            fullNameIsRequired: 'The fullName attribute is required.',
            metadataTypeIsRequired: 'The metadata type attribute is required.',
            unexpectedSmmToolingFullNameFormat: 'Unexpected source metadata member format for FullName: %s.',
            invalidResponseFromQuery: 'Invalid response from query: %s.',
            docTopicDescription: 'display help for force commands',
            docTopicLongDescription: 'Use the doc commands to display descriptions and help for commands in the force namespace.',
            docCommandsDisplayDescription: 'display help for force commands',
            docCommandsDisplayDescriptionLong: 'Displays help for commands in the force namespace.',
            docCommandsDisplayHelp: 'Displays --help output for commands in the force namespace.' +
                '\nTo display more details about the commands’ parameters, include the --json flag.',
            docCommandsListDescription: 'list the force commands',
            docCommandsListDescriptionLong: 'Lists the commands in the force namespace.',
            docCommandsListUsageDescription: 'list only docopt usage strings',
            docCommandsListUsageDescriptionLong: 'Lists docopt usage strings instead of command names and descriptions.',
            docCommandsListHelp: 'Displays a list of commands in the force namespace and their descriptions.',
            keyChainServiceCommandFailed: 'Command failed with response.\n%s',
            keyChainServiceRequired: 'Can’t get or set a keychain value without a service name.',
            keyChainAccountRequired: 'Can’t get or set a keychain value without an account name.',
            keyChainPasswordNotFound: 'Could not find password.',
            keyChainUserCanceled: 'User canceled authentication',
            keyChainCredentialParseError: 'A parse error occurred while setting a credential.',
            keychainGetCommandFailedAction: 'Determine why this command failed to get an encryption key for user %s: [%s].',
            keychainSetCommandFailedAction: 'Determine why this command failed to set an encryption key for user %s: [%s].',
            keychainPasswordNotFoundAction: 'Ensure a valid password is returned with the following command: [%s].',
            retrieveKeyChainItemFailedAction: 'Ensure that user %s has a login keychain item named %s. If not re-run authorization.',
            genericUnixKeychainInvalidPerms: 'Invalid file permissions for secret file',
            genericUnixKeychainInvalidPermsAction: 'Ensure the file %s has the file permission octal value of %s.',
            genericUnixKeychainServiceAccountMismatch: 'The service and account specified in %s do not match the version of the toolbelt.',
            genericUnixKeychainServiceAccountMismatchAction: 'Check your toolbelt version and re-auth.',
            dataExportInvalidSoql: 'Invalid SOQL query: %s',
            dataExportSoqlFailed: 'Error invoking SOQL query: %s',
            dataExportFailed: 'Export failed for soql query: %s. Error: %s.',
            dataExportQueryMalformed: 'The provided SOQL is malformed: %s',
            dataExportQueryMalformedAction: 'Check the SOQL syntax and try again.',
            dataExportRecordCount: 'Processed %s records from query: %s',
            dataExportRecordCountWarning: 'Query returned more than 200 records. Please run the command using the plan option instead.\n Record Count: %s \nQuery: %s',
            dataExportSoqlNotProvided: 'Provide a SOQL query statement or file containing a SOQL statement.',
            dataImportFileNotProvided: 'Provide a data plan or file(s).',
            dataImportFileNotFound: 'Cannot find data file. Indicate a valid path: %s.',
            dataImportFileUnknownContentType: 'Unable to determine content type for file: %s.',
            dataImportFileUnsupported: 'Content type: %s not supported.',
            dataImportTooManyFiles: 'Specify either sobjecttreefiles or a plan, but not both.',
            dataImportFileEmpty: 'Data file is empty: %s.',
            dataImportFileInvalidJson: 'data file is invalid JSON: %s',
            dataImportFileNoRefId: 'Found references in file, but no reference-id content found (%s). Was parent SObjects saved first?',
            dataImportFailed: 'Import failed from file: %s. Results: %s.',
            sourcePushFailed: 'Push failed.',
            sourcePushTimeout: 'The source push operation took longer than the specified wait time. Wait time in minutes (%s) exceeded. Increase the wait time using the --wait option, and try again.',
            sourceConflictDetected: 'Source conflict(s) detected.',
            oauthInvalidGrant: 'This org appears to have a problem with its OAuth configuration. Reason: %s \nusername: %s, \nclientId: %s, \nloginUrl: %s, \nprivateKey: %s',
            oauthInvalidGrantAction: 'Verify the OAuth configuration for this org. For JWT:' +
                `${os.EOL}Ensure the private key is correct and the cert associated with the connected app has not expired.` +
                `${os.EOL}Ensure the following OAuth scopes are configured [api, refresh_token, offline_access].` +
                `${os.EOL}Ensure the username is assigned to a profile or perm set associated with the connected app.` +
                `${os.EOL}Ensure the connected app is configured to pre-authorize admins.`,
            notSpecified: '<Not Specified>',
            warnApiVersion: 'apiVersion configuration overridden at %s'
        }
    },
    IndexErrorProcessor: {
        en_US: {
            apiMisMatch: 'The configured apiVersion %s is not supported for this org. The max apiVersion is %s',
            apiMisMatchAction: 'Run sfdx force and ensure the locally configured apiVersion is same or less than the org\'s supported max apiVersion.' +
                `${os.EOL}Run sfdx force:config:list and determine if apiVersion is overridden.` +
                `${os.EOL}Install the latest version of sfdx by running sfdx plugins:install salesforcedx@latest.`,
            server500: 'The salesforce.com servers are temporarily unable to respond to your request. We apologize for the inconvenience.' +
                `${os.EOL}Thank you for your patience, and please try again in a few moments.`,
            server500Action: 'Visit http://trust.salesforce.com for current system status and availability.'
        }
    },
    apex: {
        en_US: {
            apexTestApiReportHint: 'Run "sfdx force:apex:test:report %s" to retrieve test results.',
            apexTestApiReportForFormatHint: 'Run "sfdx force:apex:test:report %s --resultformat <format>" to retrieve test results in a different format.',
            apexTestApiReportHintWithTimeoutAction: 'Run "sfdx force:apex:test:report -i %s" to retrieve the test results. Or increase the streaming timeout by specifying the wait parameter.',
            description: 'work with Apex code',
            longDescription: 'Use the apex commands to create Apex classes, execute anonymous blocks, view your logs, run Apex tests, and view Apex test results.',
            apexCommandParamCodeCoverage: 'retrieve code coverage results',
            apexCommandParamCodeCoverageLong: 'Retrieves code coverage results.',
            apexCommandParamTestArtifactDir: 'directory to store test run files',
            apexCommandParamTestArtifactDirLong: 'Directory to store test run files.',
            apexTestCommandDescription: 'invoke Apex tests',
            apexTestCommandDescriptionLong: 'Runs Apex tests.',
            apexTestCommandHelp: 'By default, runs all Apex tests in the org’s namespace.' +
                '\nTo run specific tests, specify the class names or suite names or set a --testlevel value.' +
                '\n\nExamples:' +
                '\n   $ sfdx force:apex:test:run' +
                '\n   $ sfdx force:apex:test:run -n MyClassTest,MyOtherClassTest -r human' +
                '\n   $ sfdx force:apex:test:run -s MySuite,MyOtherSuite -c --json' +
                '\n   $ sfdx force:apex:test:run -l RunLocalTests -d <path to outputdir> -u me@my.org',
            apexTestCommandParamTestNames: 'comma-separated list of Apex test class names to execute',
            apexTestCommandParamTestNamesLong: 'Comma-separated list of Apex test class names to execute. You can’t specify both class names and suite names.',
            apexTestCommandParamTestSuites: 'comma-separated list of Apex test suite names to execute',
            apexTestCommandParamTestSuitesLong: 'Comma-separated list of Apex test suite names to execute. You can’t specify both class names and suite names.',
            apexTestCommandParamTestLevel: 'testlevel enum value',
            apexTestCommandParamTestLevelLong: 'Specifies which tests to run, using one of these TestLevel enum values:' +
                '\nRunSpecifiedTests—Only the tests that you specify are run.' +
                '\nRunLocalTests—All tests in your org are run, except the ones that originate from installed managed packages.' +
                '\nRunAllTestsInOrg—All tests are in your org and in installed managed packages are run.',
            apexTestCommandParamReporter: 'test result format emitted to stdout; --json flag overrides this parameter',
            apexTestCommandParamReporterLong: 'Format to use when displaying test results. If you also specify the --json flag, --json overrides this parameter.',
            apexTestCommandParamSynchronous: 'run tests from a single class synchronously',
            apexTestCommandParamSynchronousLong: 'Runs test methods from a single Apex class synchronously. If you don\'t specify this flag, tests are run asynchronously.',
            apexTestCommandResultFormatDeprecation: 'In salesforcedx v41 and earlier, the --resultformat parameter caused the ' +
                'apex:test:run command to wait for test results rather than finishing immediately and returning a test run ID. In salesforcedx v42 and later, the --resultformat parameter will no longer cause the command to wait. To wait for test results, include the --wait parameter.',
            apexTestCommandInvalidTestlevel: 'To use the testlevel value RunSpecifiedTests, specify either classnames or suitenames.',
            apexTestCommandInvalidSynchronousParams: 'Synchronous test runs can include test methods from only one Apex class. Omit the --synchronous flag or include tests from only one class.',
            apexTestSynchronousRunFailed: 'Unable to run tests synchronously: [%s]',
            apexReportCommandDescription: 'display test results',
            apexReportCommandDescriptionLong: 'Displays the test results for a specific test run.',
            apexReportCommandHelp: 'Displays test results for an enqueued or completed asynchronous Apex test run.' +
                '\n\nExamples:' +
                '\n   $ sfdx force:apex:test:report -i <test run id>' +
                '\n   $ sfdx force:apex:test:report -i <test run id> -r junit' +
                '\n   $ sfdx force:apex:test:report -i <test run id> -c --json',
            apexReportCommandParamTestRunId: 'ID of test run',
            apexReportCommandParamTestRunIdLong: 'The ID of test run.',
            apexReportCommandParamReporter: 'test result format emitted to stdout; --json flag overrides this parameter',
            apexReportCommandParamReporterLong: 'Format to use when displaying test results. If you also specify the --json flag, --json overrides this parameter.',
            invalidValueForSocketTimeoutHandler: 'provide a valid function for the socket timeout handler.',
            verboseDescription: 'display Apex test processing details',
            verboseLongDescription: 'Displays Apex test processing details. If JSON format is specified, processing details aren’t displayed.'
        }
    },
    apexPreCompile: {
        en_US: {
            missingOrg: 'Invalid or missing org.',
            precompileTimedOut: 'The precompile operation timed out.',
            precompileTimedOutAction: 'Increase the timeout attribute and try again.',
            precompileQueryError: 'An error occurred querying the state of the Apex cache - %s',
            precompileWarmerError: 'An error occurred running Apex pre-compilation - %s',
            invalidTimeout: 'Timeout value must be greater than zero.',
            invalidPollInterval: 'Poll interval must be greater than zero.'
        }
    },
    apexPreCompileCommand: {
        en_US: {
            invalidTimeout: 'Invalid timeout value. Value must be greater than %s minutes.',
            precompileDescription: 'how long to wait (in minutes) for Apex pre-compilation',
            precompileLongDescription: 'Specifies how long to wait (in minutes) for Apex pre-compilation to complete before running the tests or timing out.'
        }
    },
    auth: {
        en_US: {
            description: 'authorize an org for use with the Salesforce CLI',
            longDescription: 'Use the auth commands to authorize a Salesforce org for use with the Salesforce CLI.',
            clientId: 'OAuth client ID (sometimes called the consumer key)',
            clientIdLong: 'The OAuth client ID (sometimes referred to as the consumer key).',
            setDefaultDevHub: 'set the authenticated org as the default dev hub org for scratch org creation',
            setDefaultDevHubLong: 'Sets the authenticated org as the default Dev Hub org for scratch org creation.',
            setDefaultUsername: 'set the authenticated org as the default username that all commands run against',
            setDefaultUsernameLong: 'Sets the authenticated org as the default username that all commands run against.',
            setAlias: 'set an alias for the authenticated org',
            setAliasLong: 'Sets an alias for the authenticated org.',
            instanceUrl: 'the login URL of the instance the org lives on',
            instanceUrlLong: 'The login URL of the Salesforce instance that the org lives on.'
        }
    },
    auth_weblogin: {
        en_US: {
            help: 'To log in to a sandbox, set --instanceurl to https://test.salesforce.com.' +
                '\n\nExamples:' +
                '\n   $ sfdx force:auth:web:login -a TestOrg1' +
                '\n   $ sfdx force:auth:web:login -i <OAuth client id>' +
                '\n   $ sfdx force:auth:web:login -r https://test.salesforce.com',
            description: 'authorize an org using the web login flow',
            longDescription: 'Authorizes a Salesforce org by opening a browser so you can log in through salesforce.com.',
            stdin: 'OAuth client secret of personal connected app?',
            disableMasking: 'disable masking of user input (for use with problematic terminals)',
            disableMaskingLong: 'Disables masking of user input (for use with problematic terminals).'
        }
    },
    auth_jwt: {
        en_US: {
            help: 'Authorizes a Salesforce org using a private key file that has been uploaded to a personal connected app.' +
                '\n\nExamples:' +
                '\n   $ sfdx force:auth:jwt:grant -u me@my.org -f <path to jwt key file> -i <OAuth client id>' +
                '\n   $ sfdx force:auth:jwt:grant -u me@my.org -f <path to jwt key file> -i <OAuth client id> -s -a MyDefaultOrg',
            description: 'authorize an org using the JWT flow',
            longDescription: 'Authorizes a Salesforce org using the JWT flow.',
            username: 'authentication username',
            usernameLong: 'The authentication username.',
            key: 'path to a file containing the private key',
            keyLong: 'Path to a file containing the private key.'
        }
    },
    auth_sfdxurl: {
        en_US: {
            description: 'authorize an org using an SFDX auth URL',
            longDescription: 'Authorizes a Salesforce org using an SFDX auth URL.',
            help: 'Authorize a Salesforce org using an SFDX auth URL stored within a file.' +
                '\nThe file must have the format "%s" or "%s".' +
                '\nThe file must contain only the URL or be a JSON file that has a top-level property named sfdxAuthUrl.' +
                '\n\nExamples:' +
                '\n   $ sfdx force:auth:sfdxurl:store -f <path to sfdxAuthUrl file>' +
                '\n   $ sfdx force:auth:sfdxurl:store -f <path to sfdxAuthUrl file> -s -a MyDefaultOrg',
            file: 'path to a file containing the sfdx url',
            fileLong: 'Path to a file containing the SFDX URL.',
            InvalidSfdxAuthUrl: 'Invalid or missing SFDX auth URL.' +
                '\nEnsure the file exists, and that it either a) contains only the URL, or ' +
                'b) is a JSON file with a top-level property named sfdxAuthUrl.' +
                '\nEnsure the URL is in the correct format "%s" or "%s".'
        }
    },
    varargs_command: {
        en_US: {
            ArgsRequired: 'Please provide required name=value pairs for the command, quoting any values containing spaces.',
            InvalidArgsFormat: 'Setting variables must be in the format <key>=<value> or <key>="<value with spaces>" but found %s.',
            DuplicateArgs: 'Cannot set variable name \'%s\' twice for the same command.'
        }
    },
    alias: {
        en_US: {
            description: 'manage username aliases',
            longDescription: 'Use the alias commands to manage username aliases.',
            NoAliasesFound: 'Nothing to set',
            InvalidFormat: 'Setting aliases must be in the format <key>=<value> but found: [%s]'
        }
    },
    aliasSetCommand: {
        en_US: {
            help: 'You can associate an alias with only one username at a time. If you’ve set an alias multiple times, ' +
                'the alias points to the most recent username.' +
                '\n\nTo delete an alias, run "sfdx force:alias:set" with no username.' +
                '\n\nExamples:' +
                '\n   $ sfdx force:alias:set DefaultOrg=me@my.org' +
                '\n   $ sfdx force:alias:set DefaultOrg=me@my.org DevHubOrg=me@myhub.org' +
                '\n   $ sfdx force:alias:set AliasToDelete=',
            description: 'set username aliases for the Salesforce CLI',
            longDescription: 'Sets an alias that the Salesforce CLI can use for various commands and tasks.'
        }
    },
    aliasListCommand: {
        en_US: {
            help: 'Example: \n   $ sfdx force:alias:list',
            description: 'list username aliases for the Salesforce CLI',
            longDescription: 'Lists the aliases that the Salesforce CLI can use for various commands and tasks.'
        }
    },
    config: {
        en_US: {
            description: 'configure the Salesforce CLI',
            longDescription: 'Use the config commands to view and set your Salesforce CLI configuration values. ' +
                'Set your default Dev Hub and scratch org, and your default instance URL, either globally or at the project level.'
        }
    },
    sfdxConfig: {
        en_US: {
            invalidConfigValue: 'Invalid config value. %s',
            invalidApiVersion: 'Specify a valid Salesforce API version, for example, 42.0.',
            invalidInstanceUrl: 'Specify a valid Salesforce instance URL.',
            UnknownConfigKey: 'Unknown config name "%s"'
        }
    },
    configSetCommand: {
        en_US: {
            help: 'Sets the configuration variables that the Salesforce CLI uses for various commands and tasks. ' +
                'Local variables apply only to your current project. Global variables apply in any directory.' +
                '\n\nExamples:' +
                '\n   $ sfdx force:config:set defaultusername=me@my.org defaultdevhubusername=me@myhub.org' +
                '\n   $ sfdx force:config:set defaultdevhubusername=me@myhub.org -g',
            description: 'set config vars for the Salesforce CLI',
            descriptionLong: 'Sets the local and global configuration variables for the Salesforce CLI.',
            global: 'set config var globally (to be used from any directory)',
            globalLong: 'Sets the configuration variables globally, so they can be used from any directory.',
            globalHelp: 'Add the --global flag to set config outside of a project.',
            InvalidOrgValue: 'Unknown org value. Valid values are %s',
            NoConfigKeysFound: 'Please provide config name(s) and value(s) to set.',
            InvalidConfigFormat: 'Setting config vars must be in the format <key>=<value> but found %s',
            DuplicateConfigKey: 'Trying to set config name "%s" twice in the same command'
        }
    },
    configGetCommand: {
        en_US: {
            help: 'To see your default scratch org username, include "defaultusername".' +
                '\nTo see your default Dev Hub, include "defaultdevhubusername".' +
                '\nTo see your default instance URL, include "instanceUrl".' +
                '\nTo see the locations where your values are set, include the --verbose flag.' +
                '\n\nExamples:' +
                '\n   $ sfdx force:config:get defaultusername' +
                '\n   $ sfdx force:config:get defaultusername defaultdevhubusername instanceUrl' +
                '\n   $ sfdx force:config:get defaultusername defaultdevhubusername --verbose',
            description: 'get config var values for given names',
            descriptionLong: 'Gets the Salesforce CLI configuration values for your default scratch org, ' +
                'your default Dev Hub org, your default instance URL, or any combination of the three.',
            sfdxProjectValidationFailure: 'sfdx-project.json file did not validate against the schema.',
            NoConfigKeysFound: 'Please provide config name(s) to get.'
        }
    },
    configListCommand: {
        en_US: {
            help: 'Lists the config variables that the Salesforce CLI uses for various commands and tasks.',
            description: 'list config vars for the Salesforce CLI',
            descriptionLong: 'Lists the configuration variables for the Salesforce CLI.',
            key: 'Name',
            value: 'Value',
            location: 'Location'
        }
    },
    data: {
        en_US: {
            description: 'manipulate records in your org',
            longDescription: 'Use the data commands to manipulate records in your org. Commands are available to help you work with ' +
                'various APIs. Import CSV files with the Bulk API. Export and import data that includes master-detail relationships with ' +
                'the SObject Tree Save API. Perform simple CRUD operations on individual records with the REST API.',
            dataImportCommandCliDescription: 'import data into an org using SObject Tree Save API',
            dataImportCommandCliDescriptionLong: 'Imports data into an org using the SObject Tree Save API. This data can include ' +
                'master-detail relationships.',
            dataImportCommandCliHelp: 'To generate JSON files for use with force:data:tree:import, run "sfdx force:data:tree:export".' +
                '\n\nExamples:\nTo import records as individual files, first run the export commands:' +
                '\n   $ sfdx force:data:tree:export -q "SELECT Id, Name FROM Account"' +
                '\n   $ sfdx force:data:tree:export -q "SELECT Id, LastName, FirstName FROM Contact"' +
                '\nThen run the import command:' +
                '\n   $ sfdx force:data:tree:import -f Contact.json,Account.json -u me@my.org' +
                '\n\nTo import multiple data files as part of a plan, first run the export command with the -p | --plan flag:' +
                '\n   $ sfdx force:data:tree:export -p -q "SELECT Id, Name, (SELECT Id, LastName, FirstName FROM Contacts) FROM Account"' +
                '\nThen run the import command, supplying a filepath value for the -p | --plan parameter:' +
                '\n   $ sfdx force:data:tree:import -p Account-Contact-plan.json -u me@my.org',
            dataImportCommandCliFiles: 'comma-delimited, ordered paths of json files containing collection of record trees to insert',
            dataImportCommandCliFilesLong: 'Comma-delimited, ordered paths of JSON files containing a collection of record trees to insert. ' +
                'Either --sobjecttreefiles or --plan is required.',
            dataImportCommandCliContentType: 'if data file extension is not .json, provide content type (applies to all files)',
            dataImportCommandCliContentTypeLong: 'If the data file extension is not .json, provide the content type (applies to all files).',
            dataImportCommandCliPlan: 'path to plan to insert multiple data files that have master-detail relationships',
            dataImportCommandCliPlanLong: 'Path to plan to insert multiple data files that have master-detail relationships. ' +
                'Either --sobjecttreefiles or --plan is required.',
            dataImportCommandValidationFailure: 'Data plan file %s did not validate against the schema.' +
                '\nDid you run the force:data:tree:export command with the --plan flag?' +
                '\nYou can get help with the import plan schema by running $ sfdx force:data:tree:import --confighelp' +
                '\n\n%s',
            dataExportCommandCliDescription: 'export data from an org into sObject tree format for force:data:tree:import consumption',
            dataExportCommandCliDescriptionLong: 'Exports data from an org into sObject tree format for force:data:tree:import consumption.',
            dataExportCommandCliHelp: 'Generates JSON files for use with the force:data:tree:import command.' +
                '\n\nExamples:' +
                '\n   $ sfdx force:data:tree:export -q "SELECT Id, Name, (SELECT Name, Address__c FROM Properties__r) FROM Broker__c"' +
                '\n   $ sfdx force:data:tree:export -q <path to file containing soql query> -x export-demo -d /tmp/sfdx-out -p' +
                '\n\nFor more information and examples, run "sfdx force:data:tree:import -h".',
            dataExportCommandCliSoql: 'soql query, or filepath of file containing a soql query, to retrieve records',
            dataExportCommandCliSoqlLong: 'A SOQL query statement or the path of a file containing a SOQL query statement ' +
                'to retrieve the records to export.',
            dataExportCommandCliPrefix: 'prefix of generated files',
            dataExportCommandCliPrefixLong: 'Prefix of generated files.',
            dataExportCommandCliDir: 'directory to store files',
            dataExportCommandCliDirLong: 'Directory to store generated files.',
            dataExportCommandCliPlan: 'generate mulitple sobject tree files and a plan definition file for aggregated import',
            dataExportCommandCliPlanLong: 'Generates multiple sObject tree files and a plan definition file for aggregated import.'
        }
    },
    versionCommand: {
        en_US: {
            versionDescription: 'display the Salesforce API version',
            MissingVersionAttribute: 'Missing version attribute from package.json',
            UnexpectedVersionFormat: 'The version specified in package.json is unexpected: %s'
        }
    },
    mdapiConvertCommand: {
        en_US: {
            description: 'convert metadata from the Metadata API format into the Salesforce DX format',
            longDescription: 'Converts metadata retrieved via Metadata API into the source format used in Salesforce DX projects.',
            help: 'To work with metadata that you retrieved via Metadata API using the Salesforce DX tools, ' +
                'convert the metadata to the Salesforce DX source format using "sfdx force:mdapi:convert".' +
                '\n\nTo convert the source back to the Metadata API format, so that you can deploy it ' +
                'using "sfdx force:mdapi:deploy", run "sfdx force:source:convert".' +
                '\n\nExamples:' +
                '\n   $ sfdx force:mdapi:convert -r <path to source>' +
                '\n   $ sfdx force:mdapi:convert -r <path to source> -d <path to outputdir>',
            rootParam: 'the root directory containing the Metadata API–formatted metadata',
            rootParamLongDescription: 'The root directory that contains the metadata you retrieved using Metadata API.',
            outputDirectoryParam: 'the output directory to store the Salesforce DX–formatted source',
            outputDirectoryParamLongDescription: 'The directory to store your source in after it’s converted to the ' +
                'Salesforce DX format. Can be an absolute or relative path.',
            tableName: 'Converted Source',
            tableNameDups: 'Duplicates',
            dupsExplanation: 'Review any duplicate files (.dup) in the destination directory. You may need to merge the files. Otherwise, delete the unneeded one. Duplicate files are ignored by source:push/pull.'
        }
    },
    mdapiConvertApi: {
        en_US: {
            invalidPath: 'the path specified is not a directory or doesn’t contain a package.xml',
            errorProcessingPath: 'An error was encountered processing path: %s'
        }
    },
    scratchOrgApi: {
        en_US: {
            noOrgsFound: 'No scratch orgs can be found.',
            noOrgsFoundAction: 'Use one of the commands in force:auth or force:org:create to add or create new scratch orgs.'
        }
    },
    org_shape_list: {
        en_US: {
            description: 'list all org shapes you’ve created',
            longDescription: 'Lists all org shapes that you’ve created using the Salesforce CLI.',
            verbose: 'list more information about each org shape',
            verboseLong: 'Lists more information about each org shape.',
            noOrgShapes: 'No org shapes found.',
            help: 'Examples:' +
                '\n   $ sfdx force:org:shape:list' +
                '\n   $ sfdx force:org:shape:list --json' +
                '\n   $ sfdx force:org:shape:list --json > tmp/MyOrgShapeList.json'
        }
    },
    org_shape_delete: {
        en_US: {
            description: 'delete all org shapes for a target org',
            longDescription: 'Deletes all org shapes that you’ve created for an org using the Salesforce CLI.',
            verbose: 'list more information about each org shape',
            verboseLong: 'Lists more information about each org shape.',
            usernameOption: 'username for the target org',
            usernameOptionLong: 'Username for the target org.',
            targetUsernameOption: 'username for the target org',
            targetUsernameOptionLong: 'Username for the target org.',
            noPrompt: 'do not prompt for confirmation',
            noPromptLong: 'Do not prompt for confirmation.',
            noAccess: 'The org with name: %s needs to be enabled for org shape before shapes can be deleted.',
            deleteCommandYesNo: 'Delete shapes for org with name: %s?  Are you sure (y/n)?',
            humanSuccess: 'Successfully deleted shape(s) from org: %s.',
            noShapesHumanSuccess: 'No org shapes found for org: %s.',
            help: 'Examples:' +
                '\n   $ sfdx force:org:shape:delete -u me@my.org' +
                '\n   $ sfdx force:org:shape:delete -u MyOrgAlias -p' +
                '\n   $ sfdx force:org:shape:delete -u me@my.org --json' +
                '\n   $ sfdx force:org:shape:delete -u me@my.org -p --json > tmp/MyOrgShapeDelete.json'
        }
    },
    org_list: {
        en_US: {
            description: 'list all orgs you’ve created or authenticated to',
            longDescription: 'Lists all orgs that the Salesforce CLI has created or authenticated to.',
            verbose: 'list more information about each org',
            verboseLong: 'Lists more information about each org.',
            all: 'include expired, deleted, and unknown-status scratch orgs',
            allLong: 'Lists all authenticated orgs, including expired, deleted, and unknown-status scratch orgs.',
            clean: 'remove all local org authorizations for non-active orgs',
            cleanLong: 'Remove all local org authorizations for deleted or expired orgs.',
            prompt: 'Found (%s) orgs configurations to delete. Are your sure (yes/no)?',
            noPrompt: 'do not prompt for confirmation',
            noPromptLong: 'Do not prompt for confirmation.',
            noActiveScratchOrgs: 'No active scratch orgs found. Specify --all to see all scratch orgs',
            help: 'Examples:' +
                '\n   $ sfdx force:org:list' +
                '\n   $ sfdx force:org:list --verbose --json' +
                '\n   $ sfdx force:org:list --verbose --json > tmp/MyOrgList.json'
        }
    },
    package_install_get: {
        en_US: {
            cliDescription: 'retrieve the status of a package installation request',
            cliDescriptionLong: 'Retrieves the status of a package installation request.',
            help: 'Displays the status of a package install request.',
            requestId: 'ID of the package install request you want to check',
            requestIdLong: 'The ID of the package install request you want to check.',
            IN_PROGRESS: 'PackageInstallRequest is still InProgress or Unknown. You can query the status using' +
                '\nsfdx force:package:install:get -i %s -u %s',
            UNKNOWN: this.InProgress,
            SUCCESS: 'Successfully installed package [%s]'
        }
    },
    package_uninstall_get: {
        en_US: {
            cliDescription: 'retrieve status of package uninstall request',
            cliDescriptionLong: 'Retrieves the status of a package uninstall request.',
            help: 'Displays the status of a package uninstall request.',
            requestId: 'ID of the package uninstall request you want to check',
            requestIdLong: 'The ID of the package uninstall request you want to check.',
            InProgress: 'PackageUninstallRequest is still InProgress or Unknown. You can query the status using' +
                '\nsfdx force:package:uninstall:get -i %s -u %s',
            Unknown: this.InProgress,
            Success: 'Successfully uninstalled package [%s]'
        }
    },
    package1_version_create_get: {
        en_US: {
            cliDescription: 'retrieve the status of a package version creation request',
            cliDescriptionLong: 'Retrieves the status of a package version creation request.',
            help: 'Displays the status of a package upload request.',
            requestId: 'PackageUploadRequest ID',
            requestIdLong: 'The ID of the PackageUploadRequest.',
            IN_PROGRESS: 'PackageUploadRequest is still InProgress. You can query the status using' +
                '\nsfdx force:package1:version:create:get -i %s -u %s',
            SUCCESS: 'Successfully uploaded package [%s]'
        }
    },
    package_install: {
        en_US: {
            cliDescription: 'install a package in the target org',
            cliDescriptionLong: 'Installs a package in the target org.',
            help: 'Supply the ID of the package version you want to install. Installs to the default username org unless you supply the username for a different target org.' +
                '\n\nExamples:' +
                '\n   $ sfdx force:package:install -i 04t6A0000004eytQAA' +
                '\n   $ sfdx force:package:install -i 04t6A0000004eytQAA -u me@my.org',
            id: 'ID of the package version to install (starts with 04t)',
            idLong: 'The ID of the package version to install (starts with 04t).',
            wait: 'number of minutes to wait for installation status',
            waitLong: 'Maximum number of minutes to wait for installation status. The default is 0.',
            installationKey: 'installation key for key-protected package (default: null)',
            installationKeyLong: 'Installation key for installing a key-protected package. The default is null.',
            publishWait: 'number of minutes to wait for subscriber package version ID to become available in the target org',
            publishWaitLong: 'Maximum number of minutes to wait for the Subscriber Package Version ID to become available in the target org before canceling the install request. The default is 0.',
            humanSuccess: 'Successfully installed package ID %s.',
            publishWaitProgress: 'Waiting for the Subscriber Package Version ID to be published to the target org.',
            errorApvIdNotPublished: 'The package version is not fully available. If this is a recently created package' +
                ' version, please try again in a few minutes or contact the package publisher.'
        }
    },
    package_uninstall: {
        en_US: {
            cliDescription: 'uninstall a second-generation package from the target org',
            cliDescriptionLong: 'Uninstalls a second-generation package from the target org. To uninstall a first-generation package, use the Salesforce user interface.',
            help: 'Specify the package ID for a second-generation package.' +
                '\n\nExamples:' +
                '\n   $ sfdx force:package:uninstall --id 04t...' +
                '\n   $ sfdx force:package:uninstall -i 04t... -u <username of target org>' +
                '\n\nTo list the org’s installed packages, run "package:installed:list".' +
                '\n\nTo uninstall a first-generation package, from Setup, enter Installed Packages in the Quick Find box, then select Installed Packages.',
            id: 'ID of the package to uninstall (starts with 04t)',
            idLong: 'The ID of the second-generation package to uninstall (starts with 04t).',
            wait: 'number of minutes to wait for uninstall status',
            waitLong: 'Maximum number of minutes to wait for uninstall status. The default is 0.',
            defaultErrorMessage: 'Can\'t uninstall the package %s during uninstall request %s.',
            humanSuccess: 'Successfully uninstalled package ID %s.',
            action: 'Verify installed package2 ID and resolve errors, then try again.'
        }
    },
    package2: {
        en_US: {
            topicHelp: 'develop second-generation packages',
            topicHelpLong: 'Use the package2 commands to create, install, and manage second-generation packages.',
            createdLastDaysDescription: 'created in the last specified number of days (starting at 00:00:00 of first day to now; 0 for today)',
            createdLastDaysLongDescription: 'Filters the list based on the specified maximum number of days since the request was created ' +
                '(starting at 00:00:00 of first day to now; 0 for today).',
            modifiedLastDaysDescription: 'list items modified in the specified last number of days (starting at 00:00:00 of first day to now; 0 for today)',
            modifiedLastDaysLongDescription: 'Lists the items modified in the specified last number of days, starting at 00:00:00 of first day to now. Use 0 for today.',
            invalidId: 'Invalid %s: %s. Provide a valid ID that starts with %s.',
            invalidDaysNumber: 'Provide a valid positive number for %s.',
            invalidStatus: 'Invalid status \'%s\'.  Please provide one of these statuses: %s',
            package2NotEnabledAction: 'Second-generation packaging is not enabled on this org. Verify that you are authenticated to the desired org and try again. Otherwise, contact Salesforce Customer Support for more information.',
            installStatus: 'Waiting for the package install request to complete. Status = %s',
            errorMissingVersionNumber: 'The VersionNumber property must be specified.',
            errorInvalidVersionNumber: 'VersionNumber must be in the format major.minor.patch.build but the value found is [%s].',
            errorInvalidBuildNumber: 'The provided VersionNumber \'%s\' is invalid. Provide an integer value or use the keyword \'%s\' for the build number.',
            errorInvalidPatchNumber: 'The provided VersionNumber \'%s\' is not supported. Provide a patch number of 0.',
            errorInvalidMajorMinorNumber: 'The provided VersionNumber \'%s\' is invalid. Provide an integer value for the %s number.',
            errorInvalidAncestorVersionFormat: 'The ancestor versionNumber must be in the format major.minor but the value found is [%s].',
            errorNoMatchingAncestor: 'The ancestorId for ancestorVersion [%s] can\'t be found. Package 2 ID [%s].',
            errorAncestorIdVersionMismatch: 'The ancestorVersion in sfdx-project.json is not the version expected for the ancestorId you supplied. ancestorVersion [%s]. ancestorID [%s].',
            errorpackage2AncestorIdsKeyNotSupported: 'The package2AncestorIds key is no longer supported in a scratch org definition. Ancestors defined in sfdx-project.json will be included in the scratch org.'
        }
    },
    package_install_request: {
        en_US: {
            cliDescription: 'retrieve status of package install request',
            cliDescriptionLong: 'Retrieves the status of a package installion request.',
            help: 'Use force:package:install:get to display the status of a package installion request.',
            requestId: 'Package Installation Request ID',
            requestIdLong: 'ID of the package install request.',
            IN_PROGRESS: 'The package installation request is still In Progress or Unknown. You can query the status using ' +
                '\n $ sfdx force:package:install:get -i %s -u %s',
            UNKNOWN: this.InProgress,
            SUCCESS: 'Successfully installed package [%s].'
        }
    },
    package2_create: {
        en_US: {
            cliDescription: 'create a second-generation package',
            cliLongDescription: 'Creates a second-generation package.',
            help: 'First, use this command to create a second-generation package. Then create a package version.' +
                '\n\nExamples:' +
                '\n   $ sfdx force:package2:create -n PackageName -d \'My New Package\' -o Locked' +
                '\n\nRun \'sfdx force:package2:list\' to list all second-generation packages in the Dev Hub org.',
            name: 'package name',
            nameLong: 'Name of the second-generation package to create.',
            description: 'package description',
            descriptionLong: 'Description of the second-generation package.',
            noNamespace: 'creates the package with no namespace; available only for developer-controlled packages.',
            noNamespaceLong: 'Creates the package with no namespace. Available only for developer-controlled packages. ' +
                'Useful when migrating an existing org to packages, but new metadata should use a namespaced package.',
            containerOptions: '[*Managed | Unlocked | Locked] container options for the package ' +
                '(Managed=DeveloperManagedSubscriberManaged, Unlocked=DeveloperControlledSubscriberEditable, Locked=DeveloperControlledSubscriberLocked)',
            containerOptionsLong: 'Container options for the second-generation package. ' +
                '\nManaged is default. Other options include Unlocked and Locked. (Managed=DeveloperManagedSubscriberManaged, Unlocked=DeveloperControlledSubscriberEditable, Locked=DeveloperControlledSubscriberLocked). ' +
                '\nThese options determine the upgrade and editability rules.',
            humanSuccess: 'Successfully created a second-generation package (package2).'
        }
    },
    package2_update: {
        en_US: {
            cliDescription: 'update a second-generation package',
            cliLongDescription: 'Updates a second-generation package.',
            help: 'Specify a new value for each option you want to update.' +
                '\n\nExamples:' +
                '\n   $ sfdx force:package2:update --package2id 0Ho... --name \'AAnalytics\'' +
                '\n   $ sfdx force:package2:update -i 0Ho... -d \'Advanced Analytics\'' +
                '\n\nRun \'sfdx force:package2:list\' to list all second-generation packages in the Dev Hub org.',
            id: 'id of the package (starts with 0Ho)',
            idLong: 'ID of package (starts with 0Ho).',
            name: 'package name',
            nameLong: 'Name of the package to update.',
            description: 'package description',
            descriptionLong: 'Description of the package.',
            humanSuccess: 'Successfully updated the package. ID: %s.'
        }
    },
    package2_list: {
        en_US: {
            cliDescription: 'list all second-generation packages in the Dev Hub org',
            cliLongDescription: 'Lists all second-generation packages in the Dev Hub org.',
            help: 'You can view the namespace, IDs, and other details for each package.',
            namespace: 'Namespace Prefix',
            name: 'Name',
            id: 'Id',
            package2Id: 'Subscriber Package2 Id',
            description: 'Description',
            containerOptions: 'Options'
        }
    },
    package2_version_create: {
        en_US: {
            cliDescription: 'create a second-generation package version',
            cliLongDescription: 'Creates a second-generation package (package2) version in the Dev Hub org.',
            help: 'The package version is based on the package contents in the specified directory.' +
                '\n\nTo retrieve details about a package version create request, including status and package2 version ID (05i), ' +
                'run "sfdx force:package2:version:create:get -i 08c...".' +
                '\n\nTo list package version creation requests in the org, run "sfdx force:package2:version:create:list".' +
                '\n\nExamples:' +
                '\n   $ sfdx force:package2:version:create -d common' +
                '\n   $ sfdx force:package2:version:create -i 0Ho... -d common',
            package2Id: 'ID of the parent package (starts with 0Ho)',
            longPackage2Id: 'ID of parent package (starts with 0Ho).',
            dir: 'path to directory that contains the contents of the package version',
            longDir: 'The path to the directory that contains the contents of the package version.',
            branch: 'the package version’s branch',
            longBranch: 'The package version’s branch.',
            tag: 'the package version’s tag',
            longTag: 'The package version’s tag.',
            key: 'installation key for key-protected package (default: null)',
            longKey: 'Installation key for creating the key-protected package. The default is null.',
            preserve: 'temp files are preserved that would otherwise be deleted',
            longPreserve: 'Specifies that the temp files are preserved that would otherwise be deleted',
            validateschema: 'sfdx-project.json is validated against JSON schema',
            longValidateschema: 'Specifies that the sfdx-project.json file should be validated against JSON schema',
            tempFileLocation: 'The temp files are located at: %s.',
            wait: 'minutes to wait for the package version to be created',
            longWait: 'The number of minutes to wait for the package version to be created.',
            InProgress: 'Package2 version creation request is InProgress. Run "sfdx force:package2:version:create:get -i %s" to query for status.',
            Queued: 'Package2 version creation request is Queued. Run "sfdx force:package2:version:create:get -i %s" to query for status.',
            Success: 'Successfully created the package2 version [%s]. Package2 Version Id: %s. Subscriber Package2 Version Id: %s.',
            errorMissingFlags: 'Include either a %s value or a %s value. The value must match one of the packageDirectories specified in sfdx-project.json.',
            errorNoMatchingPackageDirectory: 'The %s value [%s], doesn’t match the %s value in any packageDirectories specified in sfdx-project.json.',
            errorDirectoryIdMismatch: 'The %s value, [%s], and %s value, [%s], were both found in sfdx-project.json but don’t match. If you supply both values, they must match the path and id values in one of the packageDirectories.',
            errorDependencyPair: 'Dependency must specify either a subscriberPackageVersionId or both packageId and versionNumber: %s.',
            errorNoIdInHub: 'No package2 ID was found in Dev Hub for package2 ID: %s.'
        }
    },
    package2_version_create_list: {
        en_US: {
            cliDescription: 'list package version creation requests',
            cliLongDescription: 'Lists all requests to create second-generation package (package2) versions in the Dev Hub org.',
            statusDescription: 'filter the list by version creation request status',
            statusLongDescription: 'Filters the list based on the status of version creation requests.',
            help: 'Shows the details of each request to create a package2 version that\'s run in the Dev Hub org.' +
                '\n\nAll filter parameters are applied using the AND logical operator (not OR).' +
                '\n\nTo get information about a specific request, run "sfdx force:package2:version:create:get" and supply the request ID.' +
                '\n\nExamples:' +
                '\n   $ sfdx force:package2:version:create:list' +
                '\n   $ sfdx force:package2:version:create:list --createdlastdays 3' +
                '\n   $ sfdx force:package2:version:create:list --status Error' +
                '\n   $ sfdx force:package2:version:create:list -s InProgress' +
                '\n   $ sfdx force:package2:version:create:list -c 3 -s Success',
            id: 'ID',
            status: 'Status',
            package2Id: 'Package2 Id',
            package2VersionId: 'Package2 Version Id',
            subscriberPackageVersionId: 'Subscriber Package2 Version Id',
            branch: 'Branch',
            tag: 'Tag'
        }
    },
    package2_version_create_get: {
        en_US: {
            cliDescription: 'retrieve a package version creation request',
            cliLongDescription: 'Retrieves a second-generation package version creation request in the Dev Hub org.',
            help: 'Specify the request ID for which you want to view details. ' +
                'If applicable, the command displays errors related to the request.' +
                '\n\nExamples:' +
                '\n   $ sfdx force:package2:version:create:get --package2createrequestid 08c...' +
                '\n\nTo show all requests in the org, run "sfdx package2:version:create:list".',
            requestId: 'package2 version creation request ID (starts with 08c)',
            requestIdLong: 'The ID of the package2 version creation request you want to display.',
            error: 'Error'
        }
    },
    package2_version_update: {
        en_US: {
            cliDescription: 'update a second-generation package version',
            cliLongDescription: 'Updates a second-generation package version in the Dev Hub org.',
            help: 'Specify a new value for each option you want to update.' +
                '\n\nExamples:' +
                '\n   $ sfdx force:package2:version:update --package2versionid 05i... --setasreleased' +
                '\n   $ sfdx force:package2:version:update -i 05i... -b master -t \'Release 1.0.7\'' +
                '\n\nTo display details about a package2 version, run "sfdx package2:version:get".',
            id: 'the package version ID (starts wtih 05i)',
            idLong: 'The package version ID (starts with 05i).',
            name: 'the package version name',
            nameLong: 'The package version name.',
            description: 'the package version description',
            descriptionLong: 'The second-generation package version description.',
            branch: 'the package version branch',
            branchLong: 'The second-generation package version branch.',
            tag: 'the package version tag',
            tagLong: 'The second-generation package version tag.',
            key: 'installation key for key-protected package (default: null)',
            longKey: 'Installation key for creating the key-protected package. The default is null.',
            setasreleased: 'set the package version as released (can’t be undone)',
            setasreleasedLong: 'Sets the second-generation package version as released. Second-generation packages can’t be changed to ' +
                'beta after they’ve been released.',
            humanSuccess: 'Successfully updated the package version. ID: %s.',
            previousReleased: 'To release the new package2 version, run "sfdx force:package2:version:update -s <new package2 version ID>".'
        }
    },
    package2_version_get: {
        en_US: {
            cliDescription: 'retrieve a package version in the Dev Hub org',
            cliLongDescription: 'Retrieves a package version in the Dev Hub org.',
            help: 'Examples:' +
                '\n   $ sfdx force:package2:version:get --package2versionid 05i...' +
                '\n\nTo update package version values, run "sfdx force:package2:version:update".'
        }
    },
    package2_version_list: {
        en_US: {
            cliDescription: 'list all package versions in the Dev Hub org',
            cliLongDescription: 'Lists all package2 versions in the Dev Hub org.',
            conciseDescription: 'display limited package2 version details',
            conciseLongDescription: 'Displays limited package2 version details.',
            package2IdsDescription: 'filter results on specified comma-delimited package2 ids (start with 0Ho)',
            package2IdsLongDescription: 'Filters results on the specified comma-delimited package2 IDs (start with 0Ho).',
            releasedDescription: 'display released versions only',
            releasedLongDescription: 'Displays released versions only (IsBeta=false).',
            orderByDescription: 'order by the specified package2 version fields',
            orderByLongDescription: 'Orders the list by the specified package2 version fields.',
            verboseDescription: 'display extended package2 versions detail',
            verboseLongDescription: 'Display extended package2 versions detail.',
            help: 'Displays details of each package2 version in the org.' +
                '\n\nUse --concise or --verbose to display limited or additional details, respectively.' +
                '\n\nAll filter parameters are applied using the AND logical operator (not OR).' +
                '\n\nExamples:' +
                '\n   $ sfdx force:package2:version:list --verbose --createdlastdays 3 --released --orderby PatchVersion' +
                '\n   $ sfdx force:package2:version:list --package2ids 0Ho000000000000,0Ho000000000001 --released --modifiedlastdays 0' +
                '\n   $ sfdx force:package2:version:list --released' +
                '\n   $ sfdx force:package2:version:list --concise --modifiedlastdays 0' +
                '\n   $ sfdx force:package2:version:list --concise -c 3 -r',
            name: 'Name',
            description: 'Description',
            version: 'Version',
            id: 'Package2 Version Id',
            subscriberPackageVersionId: 'Subscriber Package2 Version Id',
            package2Id: 'Package2 Id',
            package2Branch: 'Branch',
            package2Tag: 'Tag'
        }
    },
    package_installed_list: {
        en_US: {
            cliDescription: 'list the org’s installed packages',
            cliLongDescription: 'Lists all packages installed in the target org.',
            help: 'Lists all packages installed in the target org.',
            username: 'a username or alias for the target org',
            id: 'ID',
            subscriberPackageId: 'Package ID',
            subscriberPackageName: 'Package Name',
            subscriberPackageNamespace: 'Namespace',
            subscriberPackageVersionId: 'Package Version ID',
            subscriberPackageVersionName: 'Version Name',
            subscriberPackageVersionNumber: 'Version'
        }
    },
    signup: {
        en_US: {
            INVALID_ID_FIELD: 'Provide a valid template ID, in the format 0TTxxxxxxxxxxxx.',
            'T-0002': 'We couldn’t find a template with the ID specified in the scratch org definition. If you’re sure the ID is correct, contact Salesforce Support.',
            'T-0003': 'The template specified in the scratch org definition is unapproved. Contact Salesforce Support to request template approval.',
            'S-1006': 'Provide a valid email address in your scratch org definition or your %s file.',
            'S-2006': 'Provide a valid country code in your scratch org definition or your %s file.',
            'S-1017': 'Specify a namespace that’s used by a release org associated with your Dev Hub org.',
            'S-1018': 'Provide a valid My Domain value. This value can’t include double hyphens, end in a hyphen, include restricted words, or be more than 40 characters long.',
            'S-1019': 'The My Domain value you chose is already in use.',
            'S-1026': 'Provide a valid namespace value. This value must begin with a letter. It can’t include consecutive underscores, end in an underscore, be more than 15 characters long, or be a restricted or reserved namespace. Only alphanumeric characters and underscores are allowed.',
            'S-9999': 'A fatal signup error occurred. Please try again. If you still see this error, contact Salesforce Support for assistance.',
            'C-1007': 'The username provided to the org:create command is unavailable.',
            'C-1015': 'We encountered a problem while registering your My Domain value with the DNS provider. Please try again.',
            'C-1016': 'We encountered a problem while attempting to configure and approve the Connected App for your org. Verify the Connected App configuration with your Salesforce admin.',
            'C-1017': 'Provide a valid namespace prefix. This value must begin with a letter. It can’t include consecutive underscores, end in an underscore, be more than 15 characters long, or be a restricted or reserved namespace. Only alphanumeric characters and underscores are allowed.',
            'C-1020': 'We couldn’t find a template with the ID specified in the scratch org definition. If you’re sure the ID is correct, contact Salesforce Support.',
            'C-1027': 'The template specified in the Scratch Definition isn’t supported. Specify a generic edition (such as Developer or Enterprise), or specify a template ID.',
            'C-9999': 'A fatal signup error occurred. Please try again. If you still see this error, contact Salesforce Support for assistance.'
        }
    },
    generatePassword: {
        en_US: {
            description: 'generate a password for scratch org users',
            longDescription: 'Generates a password for scratch org users. Targets the usernames listed with the --onbehalfof parameter ' +
                'or the --targetusername parameter. Defaults to the defaultusername.',
            help: 'Generates and sets a random password for one or more scratch org users.' +
                `${os.EOL}${os.EOL}If you haven’t set a default Dev Hub, or if your scratch org isn’t associated with your default Dev Hub, ` +
                '--targetdevhubusername is required.' +
                `${os.EOL}${os.EOL}To see a password that was previously generated, run "sfdx force:user:display".` +
                `${os.EOL}${os.EOL}Examples:` +
                `${os.EOL}   $ sfdx force:user:password:generate` +
                `${os.EOL}   $ sfdx force:user:password:generate -u me@my.org --json` +
                `${os.EOL}   $ sfdx force:user:password:generate -o user1@my.org,user2@my.org,user3@my.org`,
            onbehalfofParam: 'comma-separated list of usernames for which to generate passwords',
            onbehalfofParamLong: 'A comma-separated list of usernames for which to generate passwords.',
            noOrgProvided: 'Please specify the org for the user(s)',
            notFoundOnDevHub: 'The scratch org does not belong to the dev hub username %s.',
            userNotFound: 'The username "%s" was not found for scratch org %s',
            usersNotFound: 'The usernames "%s" were not found for scratch org %s',
            notADevHub: 'The provided dev hub username %s is not a valid dev hub.',
            action: 'Specify the dev hub username that the scratch org belongs to with --targetdevhubusername or set a default run "sfdx force:config:set defaultdevhubusername=<username>".',
            success: 'Successfully set the password "%s" for user %s.',
            successMultiple: `Successfully set passwords:${os.EOL}`,
            viewWithCommand: 'You can see the password again by running "sfdx force:user:display -u %s".'
        }
    },
    user_create: {
        en_US: {
            description: 'create a user for a scratch org',
            longDescription: 'Creates a user for a scratch org.',
            help: 'Create a user for a scratch org, optionally setting an alias for use by the CLI, ' +
                'assigning permission sets (e.g., permsets=ps1,ps2), generating a password (e.g., generatepassword=true), ' +
                'and setting User sObject fields.' +
                `${os.EOL}${os.EOL}Examples:` +
                `${os.EOL}   $ sfdx force:user:create` +
                `${os.EOL}   $ sfdx force:user:create -a testuser1 -f config/project-user-def.json` +
                `${os.EOL}   $ sfdx force:user:create username=testuser1@my.org email=me@my.org permsets=DreamHouse` +
                `${os.EOL}   $ sfdx force:user:create -f config/project-user-def.json email=me@my.org generatepassword=true`,
            definitionfileParam: 'file path to a user definition',
            definitionfileParamLong: 'File path to a user definition.',
            setaliasParam: 'set an alias for the created username to reference within the CLI',
            setaliasParamLong: 'Sets an alias for the created username to reference within the CLI.',
            noOrgProvided: 'Please specify the org where the user should be created.',
            userNotCreated: 'User must be created before %s',
            licenseLimitExceeded: 'There are no available user licenses for the user profile "%s".',
            duplicateUsername: 'The username "%s" already exists in this or another Salesforce org. Usernames must be unique across all Salesforce orgs.',
            profileNameNotFound: 'Profile name "%s" not found in target org.  Do you need to push source?',
            success: `Successfully created user "%s" with ID [%s] for org %s.${os.EOL}You can see more details about this user by running "sfdx force:user:display -u %s".`
        }
    },
    user_list: {
        en_US: {
            description: 'lists all users of a scratch org',
            longDescription: 'Lists all users of a scratch org that the Salesforce CLI has created or authenticated.',
            help: 'The original scratch org admin is marked with "(A)"' +
                `${os.EOL}Examples:` +
                `${os.EOL}   $ sfdx force:user:list` +
                `${os.EOL}   $ sfdx force:user:list -u me@my.org --json` +
                `${os.EOL}   $ sfdx force:user:list --json > tmp/MyUserList.json`,
            noOrgProvided: 'Please specify the target org of the users to list via the --targetusername flag or setting a default username.',
            usersInOrg: 'Users in org %s'
        }
    },
    user_display: {
        en_US: {
            description: 'displays information about a user of a scratch org',
            longDescription: 'Displays information about a user of a scratch org that the Salesforce CLI has created or authenticated.',
            help: 'Output includes the profile name, org ID, access token, instance URL, login URL, and alias if applicable.' +
                `${os.EOL}Examples:` +
                `${os.EOL}   $ sfdx force:user:display` +
                `${os.EOL}   $ sfdx force:user:display -u me@my.org --json`,
            noOrgProvided: 'Please specify the user to display via the --targetusername flag or setting a default username.'
        }
    },
    source: {
        en_US: {
            description: 'sync your project with your orgs',
            longDescription: 'Use the source commands to push and pull source to and from your scratch orgs, ' +
                'to see synchronization changes between your project and scratch orgs, and to prepare your source ' +
                'for deployment to non-scratch orgs.'
        }
    },
    source_convert: {
        en_US: {
            description: 'convert Salesforce DX source into Metadata API format',
            longDescription: 'Converts source in a Salesforce DX project into metadata that you can deploy using Metadata API.',
            help: 'To convert Salesforce DX–formatted source into a format that you can deploy using Metadata API, ' +
                'run "sfdx force:source:convert". Then deploy the metadata using "sfdx force:mdapi:deploy".' +
                '\n\nTo convert Metadata API–formatted source into the Salesforce DX format, run "sfdx force:mdapi:convert".' +
                '\n\nExamples:' +
                '\n   $ sfdx force:source:convert -r <path to source>' +
                '\n   $ sfdx force:source:convert -r <path to source> -d <path to output dir>',
            rootParam: 'the source directory for the source to be converted',
            rootParamLongDescription: 'The directory that contains the Salesforce DX–formatted source to convert.',
            outputDirectoryParam: 'the output directory to export the Metadata API–formatted metadata to',
            outputDirectoryParamLongDescription: 'The output directory to export the Metadata API–formatted metadata to.',
            packageNameParam: 'the name of the package to associate with the Metadata API–formatted metadata',
            packageNameParamLongDescription: 'The name of the package to associate with the Metadata API–formatted metadata.'
        }
    },
    source_push: {
        en_US: {
            columnNumberColumn: 'Column Number',
            lineNumberColumn: 'Line Number',
            errorColumn: 'Error',
            ignoreWarningsParamDescription: 'deploy changes even if warnings are generated',
            ignoreWarningsParamLongDescription: 'Completes the deployment even if warnings are generated.'
        }
    },
    sourceConvertCommand: {
        en_US: {
            success: 'Source was successfully converted to Metadata API format and written to the location: %s',
            invalidRootDirectory: 'The package root directory does not exist: %s',
            noSourceInRootDirectory: 'The package root directory is empty.',
            rootDirectoryNotASourceDirectory: 'The package root directory is not a source directory'
        }
    },
    lightning: {
        en_US: {
            mainTopicDescriptionHelp: 'create and test Lightning component bundles',
            mainTopicLongDescriptionHelp: 'Use the lightning commands to create and test Lightning component bundles.'
        }
    },
    lightning_test: {
        en_US: {
            lightningTestCommandDescription: 'invoke Lightning component tests',
            lightningTestCommandDescriptionLong: 'Runs Lightning component tests. The Lightning Testing Service (LTS) ' +
                'unmanaged package must be installed in your org. For details, see the LTS documentation.',
            lightningTestCommandParamAppName: 'name of your Lightning test application',
            lightningTestCommandParamAppNameLong: 'Name of your Lightning test application. The name is case insensitive, ' +
                'and ".app" is optional, so "Test" and "test.app" are equivalent.' +
                '\nDefault value: Test.app',
            lightningCommandParamTestArtifactDir: 'directory path to store test run artifacts: for example, log files and test results',
            lightningCommandParamTestArtifactDirLong: 'Directory path to store test run artifacts: log files, test results, and so on.',
            lightningTestCommandParamReporter: 'test result format emitted to stdout; --json flag overrides this parameter',
            lightningTestCommandParamReporterLong: 'Format to use when displaying test results. If you also specify the --json ' +
                'flag, --json overrides this parameter.',
            lightningTestCommandParamConfig: 'path to config file for the test',
            lightningTestCommandParamConfigLong: 'Path to a test configuration file to configure WebDriver and other settings. ' +
                'For details, see the LTS documentation.',
            lightningTestCommandParamLeaveBrowserOpen: 'leave browser open',
            lightningTestCommandParamLeaveBrowserOpenLong: 'Leaves browser open after the test finishes so that you can view ' +
                'the test suite results.',
            lightningTestCommandParamTimeout: 'time (ms) to wait for results element in dom',
            lightningTestCommandParamTimeoutLong: 'Time, in milliseconds, to wait for the results element to be present in the DOM, ' +
                'before failing and moving on to the next test.',
            lightningTestCommandHelp: 'Examples:' +
                '\n   $ sfdx force:lightning:test:run' +
                '\n   $ sfdx force:lightning:test:run -a tests -r human' +
                '\n   $ sfdx force:lightning:test:run -f config/myConfigFile.json -d testResultFolder',
            lightningTestInvalidReporter: 'Unknown result format type. Must be one of the following values: %s',
            lightningTestResultRetrievalFailed: 'Unable to get test results: %s',
            lightningTestBrowserError: 'Something went wrong with the test app: %s',
            lightningTestInstallCommandDescription: 'install Lightning Testing Service unmanaged package in your org',
            lightningTestInstallCommandDescriptionLong: 'Installs a Lightning Testing Service (LTS) unmanaged package into your org.',
            lightningTestInstallCommandParamWait: 'number of minutes to wait for installation status',
            lightningTestInstallCommandParamWaitLong: 'Maximum number of minutes to wait for installation status.',
            lightningTestInstallCommandParamVersion: 'release version of Lightning Testing Service',
            lightningTestInstallCommandParamVersionLong: 'The release version of LTS unmanaged package you want to install.',
            lightningTestInstallCommandParamType: 'type of unmanaged package. \'full\' option contains both jasmine and mocha, plus examples',
            lightningTestInstallCommandParamTypeLong: 'Type of LTS unmanaged package to install. \'jasmine\' and \'mocha\' contains the essentials for development, \'full\' contains both, and adds sample components and tests in the package. \'full\' is best for \'kicking the tires\' of LTS.',
            lightningTestInstallCommandHelp: 'Examples:' +
                '\n   $ sfdx force:lightning:test:install' +
                '\n   $ sfdx force:lightning:test:install -w 0 -r v1.0' +
                '\n   $ sfdx force:lightning:test:install -t jasmine',
            scratchOrgOnly: 'Lightning Testing Service can only run in scratch orgs, because tests can permanently change org data.',
            waitNotNumber: 'The value specified for wait argument, %s, is not a number.',
            invalidType: 'The value specified for type argument, %s, is invalid.',
            invalidVersion: 'Release tag %s doesn\'t exist. For the list of available releases, go to https://github.com/forcedotcom/LightningTestingService/releases.'
        }
    },
    limits: {
        en_US: {
            mainTopicDescriptionHelp: 'view your org’s limits',
            mainTopicLongDescriptionHelp: 'Use the limits commands to view your org’s limits ' +
                'and how close you are to reaching them.'
        }
    },
    schema: {
        en_US: {
            mainTopicDescriptionHelp: 'view standard and custom objects',
            mainTopicLongDescriptionHelp: 'Use the schema commands to view the metadata for the ' +
                'standard and custom objects in your org.'
        }
    },
    visualforce: {
        en_US: {
            mainTopicDescriptionHelp: 'create and edit Visualforce files',
            mainTopicLongDescriptionHelp: 'Use the visualforce commands to create Visualforce pages and components.'
        }
    },
    project: {
        en_US: {
            mainTopicDescriptionHelp: 'set up a Salesforce DX project',
            mainTopicLongDescriptionHelp: 'Use the project commands to set up a project in the Salesforce DX format.'
        }
    },
    projectUpgrade: {
        en_US: {
            commandDescription: 'update project config files to the latest format',
            commandDescriptionLong: 'Updates project configuration and definition files to the latest format.',
            commandHelp: 'Examples:' +
                '\n $   sfdx force:project:upgrade' +
                '\n $   sfdx force:project:upgrade -f',
            forceOption: 'run all upgrades even if project has already been upgraded',
            forceOptionLong: 'Run all upgrades, even if the project definition files have already been upgraded.',
            error_validProject: 'Need a valid project to upgrade',
            error_upgradeFailed: 'Failed to upgrade: %s',
            prompt_renameProjectFile: 'Found old project file in %s. To continue upgrade, you must rename it to %s. Would you like to continue (yes/no)?',
            prompt_queuedActions: 'Found %s project update(s) to perform. Do you want to continue (list/yes/no)?',
            prompt_continue: 'Do you want to continue (yes/no)?',
            prompt_orgDefPattern: 'Check for old org definition files in the "config" directory using the pattern "config/*def.json" by entering "D" or "DEFAULT". Otherwise, enter the pattern to find your org definition files, such as "orgs/my-org-*def.json". To skip conversion of org definition files, enter "SKIP".',
            skipping: 'Skipping %s upgrades.',
            uptodate: 'Your project is up to date.',
            action_orgDefConversion: 'Upgrade %s org definition files to new format.',
            action_headsDownProject: 'Upgrade %s to headsDownCamelCase',
            action_removeUseDecomposition: 'Remove useDecomposition from %s'
        }
    },
    org_shape: {
        en_US: {
            create_shape_command_description: 'create a snapshot of org edition, features, and licenses',
            create_shape_command_description_long: 'Creates a snapshot of org edition, features, and licenses to use for scratch org creation, ' +
                'allowing your scratch org to look like another org for testing.',
            create_shape_command_help: 'Examples:' +
                '\n   $ sfdx force:org:shape:create -u me@my.org' +
                '\n   $ sfdx force:org:shape:create -u me@my.org --json --loglevel debug',
            create_shape_command_username: 'a username or alias for the target org',
            create_shape_command_username_long: 'Username or alias of the previously authorized org from which you want to create an org shape.',
            create_shape_command_no_access: 'The org needs to be enabled for org shape before one can be created.',
            create_shape_command_success: 'Successfully enqueued ShapeRepresentation creation for ID %s.'
        }
    }
};
const _getLocale = function () {
    return 'en_US';
};
module.exports = function (locale = _getLocale()) {
    return {
        getMessage(label, args, bundle = 'default') {
            const bundleLocale = messages[bundle][locale];
            if (util.isNullOrUndefined(bundleLocale)) {
                return null;
            }
            if (util.isNullOrUndefined(bundleLocale[label])) {
                throw new Error(util.format(bundleLocale.UndefinedLocalizationLabel, bundle, label, locale));
            }
            if (util.isNullOrUndefined(args)) {
                return bundleLocale[label];
            }
            else {
                const everyone = [].concat(bundleLocale[label], args);
                return util.format(...everyone);
            }
        },
        getLocale() {
            return _getLocale();
        },
        get targetusername() {
            return TARGET_USERNAME_PARAM;
        }
    };
};
