// salesforce-alm/dist/lib/orgConfigAttributes.js
/*
 * Copyright, 1999-2016, salesforce.com
 * All Rights Reserved
 * Company Confidential
 */
"use strict";
// Fields the scratch org config object is expected to have
module.exports = {
    ORG_ID: { name: 'orgId' },
    ACCESS_TOKEN: { name: 'accessToken', secret: true },
    REFRESH_TOKEN: { name: 'refreshToken', secret: true, required: false },
    INSTANCE_URL: { name: 'instanceUrl' },
    LOGIN_URL: { name: 'loginUrl', required: false },
    USERNAME: { name: 'username' },
    PASSWORD: { name: 'password', secret: true, required: false },
    CLIENT_ID: { name: 'clientId', required: false },
    PRIVATE_KEY: { name: 'privateKey', required: false },
    CLIENT_SECRET: { name: 'clientSecret', secret: true, required: false },
    CREATED_ORG_INSTANCE: { name: 'createdOrgInstance', required: false },
    CREATED: { name: 'created', required: false },
    DEV_HUB_USERNAME: { name: 'devHubUsername', required: false },
    IS_DEV_HUB: { name: 'isDevHub', required: false },
    IS_SCRATCH_ORG: { name: 'isScratchOrg', required: false },
    SCRATCH_ADMIN_USERNAME: { name: 'scratchAdminUsername', required: false },
    USER_PROFILE_NAME: { name: 'userProfileName', required: false },
    USER_ID: { name: 'userId', required: false }
};
