// salesforce-alm/dist/lib/srcStatusCommand.js
/*
 * Copyright (c) 2016, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root  or https://opensource.org/licenses/BSD-3-Clause
 */
"use strict";
const path = require('path');
const messages = require(path.join(__dirname, 'messages'));
const SrcStatusApi = require(path.join(__dirname, 'srcStatusApi'));
const syncCommandHelper = require(path.join(__dirname, 'syncCommandHelper'));
const MetadataRegistry = require(path.join(__dirname, 'metadataRegistry'));
const logger = require(path.join(__dirname, 'logApi'));
function SrcStatusCommand(context) {
    this.scratchOrg = context.org;
    this.force = this.scratchOrg.force;
    this.projectPath = this.force.config.getProjectPath();
    this.messages = messages(this.force.config.getLocale());
    this.logger = logger.child('source:status');
}
SrcStatusCommand.prototype.reject = function reject(...args) {
    const msg = this.messages.getMessage(...args);
    return Promise.reject(new Error(msg));
};
SrcStatusCommand.prototype.execute = function execute(options) {
    const rows = [];
    let api;
    return MetadataRegistry.initializeMetadataTypeInfos(this.scratchOrg)
        .then(() => {
        api = new SrcStatusApi(this.scratchOrg);
    })
        .then(() => api.doStatus(options))
        .then(() => {
        api.getLocalChanges().forEach((workspaceElement) => syncCommandHelper.createStatusLocalRows(rows, workspaceElement, this.projectPath));
        api.getRemoteChanges().forEach((workspaceElement) => syncCommandHelper.createStatusRemoteRows(rows, workspaceElement, this.projectPath));
        this.logger.styledHeader(this.logger.color.blue(this.messages.getMessage('statusCommandHumanSuccess')));
        return rows;
    });
};
/**
 * this indicated to index.js this command should produce tabular output.
 * @returns {*[]}
 */
SrcStatusCommand.prototype.getColumnData = function () {
    return syncCommandHelper.getColumnMetaInfo(this.messages);
};
module.exports = SrcStatusCommand;
