// salesforce-alm/dist/lib/staticResourceContentStrategy.js
/*
 * Copyright (c) 2017, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root  or https://opensource.org/licenses/BSD-3-Clause
 */
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const path = require('path');
const util = require('util');
const fs = require('fs-extra');
const mime = require('mime');
const srcDevUtil = require(path.join(__dirname, 'srcDevUtil'));
const staticResource_1 = require("./staticResource");
class StaticResourceContentStrategy {
    constructor(typeDef, metadataRegistry, workspaceVersion) {
        this.typeDef = typeDef;
        this.metadataRegistry = metadataRegistry;
        this.workspaceVersion = workspaceVersion;
    }
    getContentPaths(metadataFilePath) {
        const staticResource = new staticResource_1.StaticResource(metadataFilePath, this.typeDef, this.workspaceVersion);
        return staticResource.getContentPaths();
    }
    saveContent(metadataFilePath, retrievedContentFilePaths, retrievedMetadataFilePath, createDuplicates, unsupportedMimeTypes) {
        const newPaths = [];
        const deletedPaths = [];
        const staticResource = new staticResource_1.StaticResource(metadataFilePath, this.typeDef, this.workspaceVersion, retrievedMetadataFilePath, unsupportedMimeTypes);
        const [updatedPaths, duplicatePaths] = staticResource.saveResource(retrievedContentFilePaths[0], createDuplicates);
        return [newPaths, updatedPaths, deletedPaths, duplicatePaths];
    }
}
exports.StaticResourceContentStrategy = StaticResourceContentStrategy;
