// salesforce-alm/dist/lib/versionCommand.js
/*
 * Copyright (c) 2016, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root  or https://opensource.org/licenses/BSD-3-Clause
 */
"use strict";
// Node
const path = require('path');
// Thirdparty
const Promise = require('bluebird');
const optional = require('optional-js');
// Local
const Config = require(path.join(__dirname, 'configApi')).Config;
const messages = require(path.join(__dirname, 'messages'));
const srcDevUtil = require(path.join(__dirname, 'srcDevUtil'));
const _getAsciiSignature = (a) => `
                 DX DX DX
             DX DX DX DX DX DX          DX DX DX
          DX DX DX      DX DX DX    DX DX DX DX DX DX
        DX DX              DX DX DX DX DX     DX DX DX
       DX DX                 DX DX DX             DX DX    DX DX DX
      DX DX                    DX DX                DX DX DX DX DX DX DX
     DX DX                                          DX DX DX       DX DX DX
     DX DX                                            DX             DX DX DX
      DX DX                                                              DX DX
      DX DX                                                               DX DX
       DX DX                                                              DX DX
     DX DX                                                                 DX DX
   DX DX                                                                   DX DX
 DX DX                                                                     DX DX
DX DX                                                                     DX DX
DX DX                                                                    DX DX
DX DX                                                                   DX DX
 DX DX                                                    DX          DX DX
 DX DX                                                  DX DX DX DX DX DX
  DX DX                                                 DX DX DX DX DX
    DX DX DX   DX DX                     DX           DX DX
       DX DX DX DX DX                   DX DX DX DX DX DX
          DX DX  DX DX                  DX DX DX DX DX
                  DX DX              DX DX
                    DX DX DX     DX DX DX
                      DX DX DX DX DX DX                     v${a}
                          DX DX DX

* Salesforce DX Setup Guide: https://sfdc.co/sfdx_setup_guide
* Salesforce DX Developer Guide: https://sfdc.co/sfdx_dev_guide
* Salesforce CLI Command Reference: https://sfdc.co/sfdx_cli_reference
* Salesforce Extensions for VS Code: https://marketplace.visualstudio.com/items?itemName=salesforce.salesforcedx-vscode
`;
class VersionCommand {
    constructor(config) {
        this.config = optional.ofNullable(config).orElse(new Config());
        this.messages = messages(this.config.getLocale());
    }
    validate(context) {
        return Promise.resolve(srcDevUtil.fixCliContext(context));
    }
    execute() {
        return Promise.resolve({
            apiVersion: this.config.getApiVersion()
        });
    }
    getHumanSuccessMessage(obj) {
        return _getAsciiSignature(obj.apiVersion);
    }
}
module.exports = VersionCommand;
