// salesforce-alm/dist/commands/config/config_set.js
/* eslint global-require: 0 */
'use strict';
// Node
const path = require('path');
// Local
const messages = require(path.join(__dirname, '..', '..', 'lib', 'messages'))();
module.exports = function (_execCommand) {
    return {
        description: messages.getMessage('description', [], 'configSetCommand'),
        longDescription: messages.getMessage('descriptionLong', [], 'configSetCommand'),
        help: messages.getMessage('help', [], 'configSetCommand'),
        requiresProject: false,
        variableArgs: true,
        variableArgsRequired: true,
        flags: [{
                name: 'global',
                char: 'g',
                description: messages.getMessage('global', [], 'configSetCommand'),
                longDescription: messages.getMessage('globalLong', [], 'configSetCommand'),
                hasValue: false,
                required: false,
                type: 'flag'
            }],
        /**
         * the heroku cli hook
         * @param context - the cli context
         * @returns {*  }
         */
        run(context) {
            const ConfigSetCommand = require(path.join(__dirname, '..', '..', 'lib', 'config', 'ConfigSetCommand'));
            return _execCommand(new ConfigSetCommand(), context);
        }
    };
};
