// salesforce-alm/dist/commands/mdapi/mdapi_convert.js
/*
 * Copyright (c) 2016, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root  or https://opensource.org/licenses/BSD-3-Clause
 */
/* eslint global-require: 0 */
'use strict';
const path = require('path');
const messages = require(path.join(__dirname, '..', '..', 'lib', 'messages'))();
module.exports = function (_execCommand) {
    return {
        description: messages.getMessage('description', null, 'mdapiConvertCommand'),
        longDescription: messages.getMessage('longDescription', null, 'mdapiConvertCommand'),
        help: messages.getMessage('help', null, 'mdapiConvertCommand'),
        initializeMetadataRegistry: true,
        showProgress: true,
        requiresUsername: false,
        flags: [
            {
                name: 'rootdir',
                char: 'r',
                description: messages.getMessage('rootParam', null, 'mdapiConvertCommand'),
                longDescription: messages.getMessage('rootParamLongDescription', null, 'mdapiConvertCommand'),
                required: true,
                hasValue: true,
                type: 'directory'
            },
            {
                name: 'outputdir',
                char: 'd',
                description: messages.getMessage('outputDirectoryParam', null, 'mdapiConvertCommand'),
                longDescription: messages.getMessage('outputDirectoryParamLongDescription', null, 'mdapiConvertCommand'),
                required: false,
                hasValue: true,
                type: 'directory'
            }
        ],
        /**
         * the heroku cli hook
         * @param context - the cli context
         * @returns {Promise}
         */
        run(context) {
            const MdapiConvertCommand = require(path.join(__dirname, '..', '..', 'lib', 'mdapiConvertCommand'));
            return _execCommand(new MdapiConvertCommand(), context);
        }
    };
};
