// salesforce-alm/dist/commands/alias/alias_set.js
/*
 * Copyright (c) 2016, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root  or https://opensource.org/licenses/BSD-3-Clause
 */
/* eslint global-require: 0 */
'use strict';
// Node
const path = require('path');
// Local
const messages = require(path.join(__dirname, '..', '..', 'lib', 'messages'))();
module.exports = function (_execCommand) {
    return {
        description: messages.getMessage('description', [], 'aliasSetCommand'),
        longDescription: messages.getMessage('longDescription', [], 'aliasSetCommand'),
        help: messages.getMessage('help', [], 'aliasSetCommand'),
        requiresProject: false,
        variableArgs: true,
        variableArgsRequired: true,
        flags: [],
        /**
         * the heroku cli hook
         * @param context - the cli context
         * @returns {*  }
         */
        run(context) {
            const _ = require('lodash');
            const Alias = require(path.join(__dirname, '..', '..', 'lib', 'alias'));
            const logger = require(path.join(__dirname, '..', '..', 'lib', 'logApi'));
            return _execCommand({
                execute: execContext => Alias.parseAndSet(execContext.args, execContext.flags.group)
                    .then(savedAliases => {
                    logger.styledHeader(logger.color.blue('Alias Set'));
                    return _.map(savedAliases, (value, alias) => ({ alias, value }));
                }),
                getColumnData: () => [
                    { key: 'alias', label: 'Alias' },
                    { key: 'value', label: 'Value' }
                ]
            }, context);
        }
    };
};
