// salesforce-alm/dist/commands/package/package_uninstall.js
/*
 * Copyright (c) 2016, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root  or https://opensource.org/licenses/BSD-3-Clause
 */
/* eslint global-require: 0 */
'use strict';
const path = require('path');
const messages = require(path.join(__dirname, '..', '..', 'lib', 'messages'))();
const consts = require(path.join(__dirname, '..', '..', 'lib', 'constants'));
module.exports = function (_execCommand) {
    return {
        description: messages.getMessage('cliDescription', [], 'package_uninstall'),
        longDescription: messages.getMessage('cliDescriptionLong', [], 'package_uninstall'),
        help: messages.getMessage('help', [], 'package_uninstall'),
        orgType: consts.DEFAULT_USERNAME,
        supportsTargetUsername: true,
        flags: [
            {
                name: 'id',
                char: 'i',
                description: messages.getMessage('id', [], 'package_uninstall'),
                longDescription: messages.getMessage('idLong', [], 'package_uninstall'),
                hasValue: true,
                required: true,
                type: 'id'
            },
            {
                name: 'wait',
                char: 'w',
                description: messages.getMessage('wait', [], 'package_uninstall'),
                longDescription: messages.getMessage('waitLong', [], 'package_uninstall'),
                hasValue: true,
                required: false,
                type: 'minutes'
            }
        ],
        /**
         * the heroku cli hook
         * @param context - the cli context
         * @returns {Promise}
         */
        run(context) {
            const PackageUninstallCommand = require(path.join('..', '..', 'lib', 'package', 'packageUninstallCommand'));
            return _execCommand(new PackageUninstallCommand(), context);
        }
    };
};
