// salesforce-alm/dist/commands/package/package_install_get.js
/*
 * Copyright (c) 2016, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root  or https://opensource.org/licenses/BSD-3-Clause
 */
/* eslint global-require: 0 */
'use strict';
const path = require('path');
const messages = require(path.join(__dirname, '..', '..', 'lib', 'messages'))();
module.exports = function (_execCommand) {
    return {
        description: messages.getMessage('cliDescription', [], 'package_install_get'),
        longDescription: messages.getMessage('cliDescriptionLong', [], 'package_install_get'),
        help: messages.getMessage('help', [], 'package_install_get'),
        requiresProject: false,
        supportsTargetUsername: true,
        flags: [{
                name: 'requestid',
                char: 'i',
                description: messages.getMessage('requestId', [], 'package_install_get'),
                longDescription: messages.getMessage('requestIdLong', [], 'package_install_get'),
                hasValue: true,
                required: true,
                type: 'id'
            }],
        /**
         * the heroku cli hook
         * @param context - the cli context
         * @returns {Promise}
         */
        run(context) {
            const PackageInstallRequestGetCommand = require(path.join('..', '..', 'lib', 'package', 'PackageInstallRequestGetCommand'));
            return _execCommand(new PackageInstallRequestGetCommand(), context);
        }
    };
};
