// salesforce-alm/dist/lib/package/PackageInstallRequestGetCommand.js
/*
 * Copyright (c) 2016, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root  or https://opensource.org/licenses/BSD-3-Clause
 */
'use strict';
/**
 * Contains config details and operations meant for package subscriber orgs.
 * This could potentially include test orgs used by CI process for testing packages,
 * and target subscriber orgs.
 **/
const path = require('path');
const logger = require(path.join('..', 'logApi'));
const PackageInstallCommand = require(path.join(__dirname, 'packageInstallCommand'));
const DEFAULT_MAX_POLL = 0;
class PackageInstallRequestGetCommand {
    constructor() {
        this.maxRetries = DEFAULT_MAX_POLL;
        this.packageInstallRequestId = null;
        this.logger = logger.child('PackageInstallRequestGetCommand');
    }
    execute(context) {
        this.packageInstallRequestId = context.flags.requestid;
        this.packageInstallCommand = new PackageInstallCommand(context);
        return this.packageInstallCommand.poll(context, this.packageInstallRequestId, this.maxRetries);
    }
    /**
     * returns a human readable message for a cli output
     * @returns {string}
     */
    getHumanSuccessMessage(result) {
        return this.packageInstallCommand.getHumanSuccessMessage(result);
    }
}
module.exports = PackageInstallRequestGetCommand;
