// salesforce-alm/dist/commands/user/user_permset_assign.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
/*
 * Copyright, 1999-2016, salesforce.com
 * All Rights Reserved
 * Company Confidential
 */
const messages = require("../../lib/messages");
module.exports = function (_execCommand) {
    const { getMessage } = messages();
    return {
        description: getMessage('assignCommandCliDescription'),
        longDescription: getMessage('assignCommandCliDescriptionLong'),
        help: getMessage('assignCommandCliHelp'),
        requiresProject: false,
        supportsTargetUsername: true,
        flags: [{
                name: 'permsetname',
                char: 'n',
                description: getMessage('assignCommandCliName'),
                longDescription: getMessage('assignCommandCliNameLong'),
                hasValue: true,
                required: true,
                type: 'string'
            }, {
                name: 'onbehalfof',
                char: 'o',
                description: getMessage('assignCommandOnBehalfOf'),
                longDescription: getMessage('assignCommandOnBehalfOfLong'),
                hasValue: true,
                required: false,
                type: 'string',
                array: true
            }],
        run(context) {
            return __awaiter(this, void 0, void 0, function* () {
                const UserPermsetAssignCommand = require('../../lib/user/userPermsetAssignCommand').UserPermsetAssignCommand;
                return yield _execCommand(new UserPermsetAssignCommand(), context);
            });
        }
    };
};
