// salesforce-alm/dist/commands/doc/doc_commands_list.js
/*
 * Copyright (c) 2016, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root  or https://opensource.org/licenses/BSD-3-Clause
 */
/* eslint global-require: 0 */
'use strict';
const path = require('path');
const messages = require(path.join(__dirname, '..', '..', 'lib', 'messages'))();
module.exports = function (_execCommand) {
    return {
        command: 'doc:commands:list',
        description: messages.getMessage('docCommandsListDescription'),
        longDescription: messages.getMessage('docCommandsListDescriptionLong'),
        help: messages.getMessage('docCommandsListHelp'),
        requiresProject: false,
        flags: [{
                name: 'usage',
                char: 'u',
                hasValue: false,
                required: false,
                hidden: false,
                type: 'flag',
                description: messages.getMessage('docCommandsListUsageDescription'),
                longDescription: messages.getMessage('docCommandsListUsageDescriptionLong')
            }],
        run: (context) => {
            const DocCommandsListCommand = require(path.join(__dirname, '..', '..', 'lib', 'docCommandsListCommand'));
            return _execCommand(new DocCommandsListCommand(), context);
        }
    };
};
